# Data Science Aptitude Test - Daimler Group

- date: yyyy.MM.dd
- author: [name_user](https://www.linkedin.com/in/)
***

## 1. Background 

The following test consists in one scenario in which you will have to analyze and train a model for one data set. The data set contains information about year, price, transmission, mileage, fuel type and engine size of used C-class cars. The idea is training a model to **predict** which will be the price of a used *C-class car* in the market.

### Data Science Workflow stages

The competition solution workflow goes through ten stages described in the Data Science Solutions book.

1. Check Environment & libraries version.
2. Import Libraries.
3. Question or problem definition.
4. Acquire dataset or training and testing data.
5. Analyze, identify patterns, and explore the data.
6. Wrangle, prepare, cleanse the data.
7. Feature Engineering.
8. Model, predict and MLOps.
9. Visualize, report, and present the problem solving steps and final solution.
10. Create or solve AI pipeline on Cloud, architecture solution and shows results.

Could be to exist another substeps during Data Processing and Data Modeling, for to train and deploy many machine learning and/or deep learning combination pipelines.

## 2. This notebook covers only:

1. [X] Upload the csv file to your workspace and load it into a data frame.
2. [X] Look for null values and outliers. Remove, keep or impute them and explain why you did so.
3. [X] Show the main statistics (mean, standard deviation…) of the numerical columns of the data set. Are any of the variables skewed? (You can use any visualization you need to answer the last question).
4.	Train a model for the prediction of the price. Explain why you chose the model that you have trained.
5.	Test the model and obtain some performance metrics from it. Would you say that the model has a good performance? Why?
6.	Would you say that you have enough information to predict the price of an EQC (electric C-class)? Why?

## 3. The business problem

To be able to predict used cars market value can help both buyers and sellers.

Used car sellers (dealers): They are one of the biggest target group that can be interested in results of this study. If used car sellers better understand what makes a car desirable, what the important features are for a used car, then they may consider this knowledge and offer a better service.

Online pricing services: There are websites that offers an estimate value of a car. They may have a good prediction model. However, having a second model may help them to give a better prediction to their users. Therefore, the model developed in this study may help online web services that tells a used car’s market value.

Individuals: There are lots of individuals who are interested in the used car market at some points in their life because they wanted to sell their car or buy a used car. In this process, it’s a big corner to pay too much or sell less then it’s market value.

What determines the price of used cars? 

<img src="https://s1.dmcdn.net/v/L5C9Y1QCcGAoct-M5/x1080" width="200">


The value of a car drops right from the moment it is bought and the depreciation continues with each passing year. In fact, in the first year itself, the value of a car decreases by 20 percent of its initial value. The make and model of a car, total kilometers driven, overall condition of the vehicle and various other factors further affect the car’s resale value .

Our results show that Random Forest model and K-Means clustering with linear regression yield the best results, but are compute heavy.
Conventional linear regression also yielded satisfactory results, with the advantage of a significantly lower training time in comparison to the aforementioned methods. We check XGBoost, Linear Regression and Decision Tree estimators.

## 4. ETL (Extract Transform Load)

Loading dataset from csv file. Anyway in other notebook in attachment, we will be able to scrape specialized website and get other important features.

In [0]:
dbutils.fs.rm ("/FileStore/tables/cclass-1.csv")

Out[9]: True

In [0]:
dbutils.fs.ls ("/FileStore/tables")

Out[10]: [FileInfo(path='dbfs:/FileStore/tables/cclass.csv', name='cclass.csv', size=180900, modificationTime=1625648272000),
 FileInfo(path='dbfs:/FileStore/tables/portoseguros/', name='portoseguros/', size=0, modificationTime=0),
 FileInfo(path='dbfs:/FileStore/tables/train-1.csv', name='train-1.csv', size=17333449, modificationTime=1621949479000),
 FileInfo(path='dbfs:/FileStore/tables/train-2.csv', name='train-2.csv', size=17333449, modificationTime=1621949556000),
 FileInfo(path='dbfs:/FileStore/tables/train.csv', name='train.csv', size=17333449, modificationTime=1621949388000),
 FileInfo(path='dbfs:/FileStore/tables/trainCleansing.csv', name='trainCleansing.csv', size=119644256, modificationTime=1622649888000),
 FileInfo(path='dbfs:/FileStore/tables/train_csv.zip', name='train_csv.zip', size=3208505, modificationTime=1621949339000)]

In [0]:
# File location and type cclass dataset stored on Databricks
file_location = "/FileStore/tables/cclass.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
    .option("inferschema", infer_schema) \
    .option("header", first_row_is_header) \
    .option("sep", delimiter) \
    .load(file_location)

display(df)

model year price transmission mileage fuelType engineSize C Class 2020 30495 Automatic 1200 Diesel 2.0 C Class 2020 29989 Automatic 1000 Petrol 1.5 C Class 2020 37899 Automatic 500 Diesel 2.0 C Class 2019 30399 Automatic 5000 Diesel 2.0 C Class 2019 29899 Automatic 4500 Diesel 2.0 C Class 2020 30999 Automatic null Diesel 2.0 C Class 2020 35999 Automatic 500 Diesel 2.0 C Class 2019 37990 Automatic 1412 Petrol 3.0 C Class 2019 28990 Automatic 3569 Diesel 2.0 C Class 2019 28990 Automatic 3635 Diesel 2.0 C Class 2013 9995 Automatic 44900 Petrol 1.6 C Class 2012 6995 Automatic 88200 Diesel 2.1 C Class 2012 7495 Automatic 115000 Diesel 2.1 C Class 2011 8995 Automatic 69250 Diesel 2.1 C Class 2015 14995 Automatic 49850 Diesel 2.1 C Class 2013 8595 Automatic 82685 Diesel 2.1 C Class 2016 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013 10977 Automatic 72000 Diesel 2.1 C Class 2016 17977 Automatic 40000 Diesel 1.6 C Class 2015 12790 Automatic 52000 Diesel 2.1 C Class 2017 14488 Automatic 66472 Diesel 2.1 C Class 2011 7500 Manual 91000 Diesel 2.1 C Class 2013 9750 Automatic 69700 Diesel 2.1 C Class 2015 15995 Automatic 39000 Diesel 1.6 C Class 2016 15675 Manual 39287 Diesel 2.1 C Class 2016 16475 Manual 40569 Diesel 2.1 C Class 2016 15975 Manual 37419 Petrol 2.0 C Class 2016 16475 Automatic 34452 Diesel 2.1 C Class 2015 15975 Automatic 41567 Hybrid 2.1 C Class 2015 16495 Automatic null Hybrid 2.0 C Class 2017 17795 Automatic 37778 Diesel 2.1 C Class 2017 17875 Automatic 44325 Diesel 2.1 C Class 2016 14000 Automatic 45000 Diesel 2.1 C Class 2017 16500 Automatic 75000 Diesel 1.6 C Class 2016 21000 Automatic 36000 Diesel 2.1 C Class 2015 15500 Automatic 64000 Diesel 2.1 C Class 2013 7250 Manual 106000 Diesel 2.1 C Class 2015 13000 Automatic 67150 Diesel 2.1 C Class 2012 11000 Automatic 43426 Petrol 1.6 C Class 2016 13300 Manual 42622 Diesel 2.1 C Class 2016 16000 Manual 34903 Diesel 2.1 C Class 2014 12500 Automatic 35345 Diesel 2.1 C Class 2015 14495 Manual 23146 Petrol 2.0 C Class 2017 24500 Automatic 16000 Diesel 2.1 C Class 2013 23990 Automatic 44000 Petrol 6.2 C Class 2015 17250 Automatic 30000 Diesel 2.1 C Class 2015 14200 Automatic 22000 Petrol 1.6 C Class 2017 22950 Automatic 19000 Petrol 2.0 C Class 2013 10500 Automatic 90000 Diesel 2.1 C Class 2014 15650 Automatic 32000 Diesel 2.1 C Class 2015 14850 Automatic 90000 Diesel 2.1 C Class 2017 22400 Automatic 37000 Diesel 2.1 C Class 2017 19950 Automatic 23000 Diesel 2.1 C Class 2014 9990 Automatic 78000 Diesel 2.1 C Class 2015 11990 Automatic 47000 Diesel 2.1 C Class 2017 20990 Automatic 18000 Diesel 2.1 C Class 2015 14490 Automatic 64000 Diesel 2.1 C Class 2013 10490 Automatic 50000 Diesel 2.1 C Class 2016 15990 Automatic 65000 Diesel 2.1 C Class 2016 17850 Automatic 38000 Diesel 2.1 C Class 2016 16450 Automatic 70000 Diesel 2.1 C Class 2013 11350 Automatic 69000 Diesel 2.1 C Class 2014 12900 Automatic 41000 Petrol 1.6 C Class 2018 26000 Automatic 16250 Petrol 1.5 C Class 2017 22767 Automatic 30676 Diesel 2.1 C Class 2015 12000 Automatic 40005 Diesel 2.1 C Class 2017 18999 Automatic 38000 Hybrid 2.0 C Class 2016 14000 Automatic 45000 Diesel 2.1 C Class 2015 12000 Automatic 40005 Diesel 2.1 C Class 2009 4343 Automatic 120000 Diesel 2.1 C Class 2016 16500 Automatic 51936 Diesel 2.1 C Class 2019 27750 Automatic 4000 Diesel 2.0 C Class 2019 28950 Automatic 8500 Diesel 2.0 C Class 2007 3495 Automatic 119168 Petrol 1.8 C Class 2017 59740 Automatic 21800 Petrol 4.0 C Class 2019 44990 Automatic 19999 Petrol 4.0 C Class 2019 54990 Automatic 1600 Petrol 4.0 C Class 2015 15995 Automatic 46221 Diesel 2.1 C Class 2016 14000 Automatic 45000 Diesel 2.1 C Class 2015 12000 Automatic 40005 Diesel 2.1 C Class 2017 22767 Automatic 30676 Diesel 2.1 C Class 2014 12998 Semi-Auto 5271

In [0]:
# Create a custom schema to pass data types correctly.
from pyspark.sql.types import *

customSchema = StructType([ \
 StructField('model', StringType(), True), \
 StructField('year', DateType(), True), \
 StructField('price', IntegerType(), True), \
 StructField('transmission', StringType(), True), \
 StructField('mileage', IntegerType(), True), \
 StructField('fuelType', StringType(), True), \
 StructField('engineSize', StringType(), True)
                          ])

In [0]:
# Load dataset with custom schema
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
    .option("header", first_row_is_header) \
    .option("sep", delimiter) \
    .load(file_location, schema=customSchema)

display(df)

model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 29989 Automatic 1000 Petrol 1.5 C Class 2020-01-01 37899 Automatic 500 Diesel 2.0 C Class 2019-01-01 30399 Automatic 5000 Diesel 2.0 C Class 2019-01-01 29899 Automatic 4500 Diesel 2.0 C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2020-01-01 35999 Automatic 500 Diesel 2.0 C Class 2019-01-01 37990 Automatic 1412 Petrol 3.0 C Class 2019-01-01 28990 Automatic 3569 Diesel 2.0 C Class 2019-01-01 28990 Automatic 3635 Diesel 2.0 C Class 2013-01-01 9995 Automatic 44900 Petrol 1.6 C Class 2012-01-01 6995 Automatic 88200 Diesel 2.1 C Class 2012-01-01 7495 Automatic 115000 Diesel 2.1 C Class 2011-01-01 8995 Automatic 69250 Diesel 2.1 C Class 2015-01-01 14995 Automatic 49850 Diesel 2.1 C Class 2013-01-01 8595 Automatic 82685 Diesel 2.1 C Class 2016-01-01 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017-01-01 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018-01-01 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016-01-01 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014-01-01 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013-01-01 10977 Automatic 72000 Diesel 2.1 C Class 2016-01-01 17977 Automatic 40000 Diesel 1.6 C Class 2015-01-01 12790 Automatic 52000 Diesel 2.1 C Class 2017-01-01 14488 Automatic 66472 Diesel 2.1 C Class 2011-01-01 7500 Manual 91000 Diesel 2.1 C Class 2013-01-01 9750 Automatic 69700 Diesel 2.1 C Class 2015-01-01 15995 Automatic 39000 Diesel 1.6 C Class 2016-01-01 15675 Manual 39287 Diesel 2.1 C Class 2016-01-01 16475 Manual 40569 Diesel 2.1 C Class 2016-01-01 15975 Manual 37419 Petrol 2.0 C Class 2016-01-01 16475 Automatic 34452 Diesel 2.1 C Class 2015-01-01 15975 Automatic 41567 Hybrid 2.1 C Class 2015-01-01 16495 Automatic null Hybrid 2.0 C Class 2017-01-01 17795 Automatic 37778 Diesel 2.1 C Class 2017-01-01 17875 Automatic 44325 Diesel 2.1 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2017-01-01 16500 Automatic 75000 Diesel 1.6 C Class 2016-01-01 21000 Automatic 36000 Diesel 2.1 C Class 2015-01-01 15500 Automatic 64000 Diesel 2.1 C Class 2013-01-01 7250 Manual 106000 Diesel 2.1 C Class 2015-01-01 13000 Automatic 67150 Diesel 2.1 C Class 2012-01-01 11000 Automatic 43426 Petrol 1.6 C Class 2016-01-01 13300 Manual 42622 Diesel 2.1 C Class 2016-01-01 16000 Manual 34903 Diesel 2.1 C Class 2014-01-01 12500 Automatic 35345 Diesel 2.1 C Class 2015-01-01 14495 Manual 23146 Petrol 2.0 C Class 2017-01-01 24500 Automatic 16000 Diesel 2.1 C Class 2013-01-01 23990 Automatic 44000 Petrol 6.2 C Class 2015-01-01 17250 Automatic 30000 Diesel 2.1 C Class 2015-01-01 14200 Automatic 22000 Petrol 1.6 C Class 2017-01-01 22950 Automatic 19000 Petrol 2.0 C Class 2013-01-01 10500 Automatic 90000 Diesel 2.1 C Class 2014-01-01 15650 Automatic 32000 Diesel 2.1 C Class 2015-01-01 14850 Automatic 90000 Diesel 2.1 C Class 2017-01-01 22400 Automatic 37000 Diesel 2.1 C Class 2017-01-01 19950 Automatic 23000 Diesel 2.1 C Class 2014-01-01 9990 Automatic 78000 Diesel 2.1 C Class 2015-01-01 11990 Automatic 47000 Diesel 2.1 C Class 2017-01-01 20990 Automatic 18000 Diesel 2.1 C Class 2015-01-01 14490 Automatic 64000 Diesel 2.1 C Class 2013-01-01 10490 Automatic 50000 Diesel 2.1 C Class 2016-01-01 15990 Automatic 65000 Diesel 2.1 C Class 2016-01-01 17850 Automatic 38000 Diesel 2.1 C Class 2016-01-01 16450 Automatic 70000 Diesel 2.1 C Class 2013-01-01 11350 Automatic 69000 Diesel 2.1 C Class 2014-01-01 12900 Automatic 41000 Petrol 1.6 C Class 2018-01-01 26000 Automatic 16250 Petrol 1.5 C Class 2017-01-01 22767 Automatic 30676 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2017-01-01 18999 Automatic 38000 Hybrid 2.0 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2009-01-01 4343 Automatic 120000 Diesel 2.1 C Class 2016-01-01 16500 Automatic 51936 Diesel 2.1 C Class 2019-01-01 27750 Automatic 4000 

In [0]:
# Check basic statistics like mean, standard deviation, min and max
display(df.describe(["price","mileage"]))

summary price mileage count 3899 3892 mean 23674.28699666581 22396.172661870503 stddev 8960.21821842348 22641.50657329362 min 1290 1 max 88995 173000

> Check distribution later.

In [0]:
# Create a view or temporary table to manipulate in SparkSQL
temp_table_name = "cclass_csv"

df.createOrReplaceTempView(temp_table_name)


In [0]:
%sql

/* Query the created temp table in a SQL cell, show all features limited 25 rows only */
SELECT * FROM `cclass_csv` LIMIT 25;


model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 29989 Automatic 1000 Petrol 1.5 C Class 2020-01-01 37899 Automatic 500 Diesel 2.0 C Class 2019-01-01 30399 Automatic 5000 Diesel 2.0 C Class 2019-01-01 29899 Automatic 4500 Diesel 2.0 C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2020-01-01 35999 Automatic 500 Diesel 2.0 C Class 2019-01-01 37990 Automatic 1412 Petrol 3.0 C Class 2019-01-01 28990 Automatic 3569 Diesel 2.0 C Class 2019-01-01 28990 Automatic 3635 Diesel 2.0 C Class 2013-01-01 9995 Automatic 44900 Petrol 1.6 C Class 2012-01-01 6995 Automatic 88200 Diesel 2.1 C Class 2012-01-01 7495 Automatic 115000 Diesel 2.1 C Class 2011-01-01 8995 Automatic 69250 Diesel 2.1 C Class 2015-01-01 14995 Automatic 49850 Diesel 2.1 C Class 2013-01-01 8595 Automatic 82685 Diesel 2.1 C Class 2016-01-01 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017-01-01 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018-01-01 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016-01-01 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014-01-01 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013-01-01 10977 Automatic 72000 Diesel 2.1 C Class 2016-01-01 17977 Automatic 40000 Diesel 1.6 C Class 2015-01-01 12790 Automatic 52000 Diesel 2.1

## 5. Data Visualization

The Exploratory Data Analysis (EDA)

While exploring the data, we will look at the different combinations of features with the help of visuals. This will help us to understand our data better and give us some clue about pattern in data.

In [0]:
# Validating data
df.printSchema()

root
 |-- model: string (nullable = true)
 |-- year: date (nullable = true)
 |-- price: integer (nullable = true)
 |-- transmission: string (nullable = true)
 |-- mileage: integer (nullable = true)
 |-- fuelType: string (nullable = true)
 |-- engineSize: string (nullable = true)



In [0]:
# Dimension of Dataset
type(df)
# No existe shape en pyspark.sql.dataframe
print(f"El dataset tiene {df.count()} observaciones y {len(df.columns)} atributos.")

El dataset tiene 3899 observaciones y 7 atributos.


In [0]:
df.sample(False, 0.5, 17).show(10)

+-------+----------+-----+------------+-------+--------+----------+
|  model|      year|price|transmission|mileage|fuelType|engineSize|
+-------+----------+-----+------------+-------+--------+----------+
|C Class|2020-01-01|29989|   Automatic|   1000|  Petrol|       1.5|
|C Class|2020-01-01|37899|   Automatic|    500|  Diesel|       2.0|
|C Class|2019-01-01|29899|   Automatic|   4500|  Diesel|       2.0|
|C Class|2019-01-01|28990|   Automatic|   3569|  Diesel|       2.0|
|C Class|2011-01-01| 8995|   Automatic|  69250|  Diesel|       2.1|
|C Class|2015-01-01|14995|   Automatic|  49850|  Diesel|       2.1|
|C Class|2017-01-01|23500|   Semi-Auto|  17000|  Diesel|       2.1|
|C Class|2016-01-01|17977|   Automatic|  40000|  Diesel|       1.6|
|C Class|2015-01-01|12790|   Automatic|  52000|  Diesel|       2.1|
|C Class|2017-01-01|14488|   Automatic|  66472|  Diesel|       2.1|
+-------+----------+-----+------------+-------+--------+----------+
only showing top 10 rows



In [0]:
# Show 25 sample values
df.sample(False, 0.5, 17).toPandas().sample(25)

model 
 year 
 price 
 transmission 
 mileage 
 fuelType 
 engineSize 
 
 
 
 
 189 
 C Class 
 2019-01-01 
 22455 
 Automatic 
 3761.0 
 Petrol 
 1.5 
 
 
 205 
 C Class 
 2014-01-01 
 15971 
 Semi-Auto 
 14354.0 
 Petrol 
 2.0 
 
 
 1260 
 C Class 
 2019-01-01 
 28399 
 Semi-Auto 
 3061.0 
 Petrol 
 1.5 
 
 
 1069 
 C Class 
 2019-01-01 
 26599 
 Semi-Auto 
 12961.0 
 Petrol 
 1.5 
 
 
 1234 
 C Class 
 2019-01-01 
 67199 
 Semi-Auto 
 2270.0 
 Petrol 
 4.0 
 
 
 1161 
 C Class 
 2019-01-01 
 29500 
 Semi-Auto 
 7610.0 
 Diesel 
 2.0 
 
 
 1857 
 C Class 
 2015-01-01 
 17500 
 Automatic 
 45566.0 
 Diesel 
 2.1 
 
 
 1659 
 C Class 
 2011-01-01 
 7999 
 Manual 
 64000.0 
 Petrol 
 1.8 
 
 
 1048 
 C Class 
 2019-01-01 
 30099 
 Semi-Auto 
 4701.0 
 Diesel 
 1.6 
 
 
 1061 
 C Class 
 2019-01-01 
 25699 
 Semi-Auto 
 11793.0 
 Petrol 
 1.5 
 
 
 460 
 C Class 
 2019-01-01 
 31499 
 Automatic 
 2206.0 
 Diesel 
 2.0 
 
 
 1893 
 C Class 
 2019-01-01 
 24790 
 Automatic 
 10637.0 
 Diesel 
 2.0 
 
 
 1245 
 C Class 
 2019-01-01 
 27699 
 Semi-Auto 
 5426.0 
 Petrol 
 1.5 
 
 
 1578 
 C Class 
 2016-01-01 
 15600 
 Manual 
 25997.0 
 Diesel 
 1.6 
 
 
 1708 
 C Class 
 2017-01-01 
 17599 
 Automatic 
 50212.0 
 Diesel 
 1.6 
 
 
 1139 
 C Class 
 2019-01-01 
 28949 
 Semi-Auto 
 6565.0 
 Diesel 
 2.0 
 
 
 824 
 C Class 
 2019-01-01 
 29995 
 Semi-Auto 
 1.0 
 Diesel 
 2.0 
 
 
 1677 
 C Class 
 2015-01-01 
 11500 
 Automatic 
 101540.0 
 Diesel 
 2.1 
 
 
 1602 
 C Class 
 2016-01-01 
 18600 
 Automatic 
 24836.0 
 Diesel 
 1.6 
 
 
 629 
 C Class 
 2019-01-01 
 29495 
 Automatic 
 4999.0 
 Diesel 
 2.0 
 
 
 257 
 C Class 
 2015-01-01 
 15790 
 Semi-Auto 
 19830.0 
 Petrol 
 2.0 
 
 
 1604 
 C Class 
 2019-01-01 
 23400 
 Automatic 
 8000.0 
 Petrol 
 1.5 
 
 
 410 
 C Class 
 2019-01-01 
 20480 
 Semi-Auto 
 14620.0 
 Petrol 
 1.5 
 
 
 1524 
 C Class 
 2017-01-01 
 15999 
 Automatic 
 24518.0 
 Diesel 
 2.1 
 
 
 868 
 C Class 
 2017-01-01 
 26949 
 Automatic 
 28058.0 
 Petrol 
 2.0

In [0]:
%sql
SELECT price, fuelType
FROM cclass_csv

price fuelType 30495 Diesel 29989 Petrol 37899 Diesel 30399 Diesel 29899 Diesel 30999 Diesel 35999 Diesel 37990 Petrol 28990 Diesel 28990 Diesel 9995 Petrol 6995 Diesel 7495 Diesel 8995 Diesel 14995 Diesel 8595 Diesel 19250 Diesel 23500 Diesel 24995 Diesel 29995 Petrol 17990 Petrol 14477 Diesel 10977 Diesel 17977 Diesel 12790 Diesel 14488 Diesel 7500 Diesel 9750 Diesel 15995 Diesel 15675 Diesel 16475 Diesel 15975 Petrol 16475 Diesel 15975 Hybrid 16495 Hybrid 17795 Diesel 17875 Diesel 14000 Diesel 16500 Diesel 21000 Diesel 15500 Diesel 7250 Diesel 13000 Diesel 11000 Petrol 13300 Diesel 16000 Diesel 12500 Diesel 14495 Petrol 24500 Diesel 23990 Petrol 17250 Diesel 14200 Petrol 22950 Petrol 10500 Diesel 15650 Diesel 14850 Diesel 22400 Diesel 19950 Diesel 9990 Diesel 11990 Diesel 20990 Diesel 14490 Diesel 10490 Diesel 15990 Diesel 17850 Diesel 16450 Diesel 11350 Diesel 12900 Petrol 26000 Petrol 22767 Diesel 12000 Diesel 18999 Hybrid 14000 Diesel 12000 Diesel 4343 Diesel 16500 Diesel 27750 Diesel 28950 Diesel 3495 Petrol 59740 Petrol 44990 Petrol 54990 Petrol 15995 Diesel 14000 Diesel 12000 Diesel 22767 Diesel 12998 Diesel 17998 Diesel 28998 Petrol 24998 Petrol 19498 Diesel 23498 Diesel 25498 Diesel 24998 Diesel 18498 Hybrid 15498 Diesel 20995 Diesel 31995 Petrol 23995 Diesel 16995 Diesel 24995 Petrol 26998 Petrol 20498 Diesel 27498 Diesel 16998 Diesel 27498 Diesel 27698 Diesel 27498 Petrol 32498 Diesel 24498 Diesel 21498 Diesel 18998 Diesel 15998 Diesel 9298 Diesel 22998 Diesel 12998 Diesel 22998 Diesel 17998 Diesel 28498 Diesel 18998 Diesel 20998 Diesel 21998 Diesel 13998 Diesel 24498 Diesel 20998 Hybrid 30498 Diesel 19498 Diesel 15298 Diesel 33998 Petrol 32998 Petrol 16498 Diesel 15498 Diesel 15498 Diesel 18998 Diesel 24495 Diesel 18990 Diesel 10995 Diesel 15698 Diesel 22498 Petrol 19298 Diesel 16498 Hybrid 19998 Petrol 20685 Diesel 20489 Diesel 13800 Diesel 31998 Petrol 29998 Petrol 24998 Petrol 32998 Petrol 32998 Petrol 29998 Petrol 23998 Diesel 20998 Hybrid 16498 Diesel 11998 Diesel 12495 Diesel 19298 Petrol 23998 Diesel 12495 Diesel 17498 Hybrid 16198 Diesel 18498 Diesel 24998 Petrol 24998 Petrol 21998 Petrol 29498 Diesel 29498 Diesel 16998 Diesel 33998 Petrol 32998 Petrol 18998 Diesel 32498 Petrol 25798 Diesel 20498 Diesel 18498 Diesel 14498 Petrol 29498 Diesel 17998 Petrol 13498 Diesel 19698 Hybrid 22498 Hybrid 11998 Diesel 11998 Petrol 39998 Hybrid 23498 Diesel 57998 Petrol 16315 Diesel 13991 Diesel 16998 Diesel 16498 Diesel 16998 Diesel 13690 Diesel 14490 Diesel 17998 Petrol 29998 Diesel 18998 Hybrid 18998 Petrol 28998 Diesel 24998 Petrol 27998 Diesel 16798 Diesel 19998 Diesel 29950 Diesel 20790 Hybrid 9490 Petrol 15690 Diesel 14990 Diesel 17290 Diesel 19490 Diesel 28990 Diesel 24990 Diesel 28990 Petrol 16298 Diesel 14698 Diesel 13998 Petrol 14698 Diesel 10498 Diesel 11998 Diesel 15698 Petrol 16998 Diesel 17998 Hybrid 20998 Diesel 13691 Diesel 13491 Diesel 21998 Petrol 14500 Diesel 15990 Diesel 20490 Petrol 20890 Petrol 14490 Diesel 21990 Petrol 14750 Diesel 44990 Petrol 24690 Petrol 18270 Diesel 26625 Diesel 13361 Petrol 17490 Diesel 26427 Diesel 29290 Diesel 23047 Diesel 13621 Diesel 18200 Diesel 16991 Diesel 23498 Petrol 14698 Diesel 9798 Diesel 17998 Diesel 18998 Diesel 20700 Diesel 16990 Diesel 11791 Diesel 18191 Diesel 20991 Diesel 13321 Diesel 21998 Diesel 15998 Hybrid 13498 Diesel 20871 Diesel 25891 Petrol 20591 Diesel 19991 Diesel 15990 Diesel 30055 Diesel 21650 Petrol 25990 Petrol 17690 Diesel 30950 Diesel 24950 Petrol 14998 Diesel 19798 Hybrid 24498 Petrol 19991 Diesel 21991 Diesel 22591 Diesel 18790 Diesel 25950 Petrol 29950 Diesel 14990 Diesel 31950 Diesel 20990 Diesel 28990 Petrol 27000 Diesel 12590 Diesel 20885 Diesel 25885 Petrol 14965 Diesel 16595 Diesel 17940 Diesel 19940 Diesel 16498 Diesel 14500 Diesel 27791 Diesel 26499 Diesel 31990 Diesel 20999 Petrol 31999 Diesel 17495 Petrol 25495 Petrol 36995 Diesel 30999 Diesel 29999 Diesel 17880 Diesel 13660 Diesel 23920 Diesel 22960 Diesel 16590 Diesel 17990 Di

> First part of multivariate analysis, price is **right skewed** for most of fuelType feature, most frequent for Diesel and Petrol. Right-skewed distributions are also called positive-skew distributions. That’s because there is a long tail in the positive direction on the number line. The mean is also to the right of the peak.

In [0]:
%sql
SELECT mileage, fuelType FROM cclass_csv

mileage fuelType 1200 Diesel 1000 Petrol 500 Diesel 5000 Diesel 4500 Diesel null Diesel 500 Diesel 1412 Petrol 3569 Diesel 3635 Diesel 44900 Petrol 88200 Diesel 115000 Diesel 69250 Diesel 49850 Diesel 82685 Diesel 32506 Diesel 17000 Diesel 17744 Diesel 64291 Petrol 41011 Petrol 72000 Diesel 72000 Diesel 40000 Diesel 52000 Diesel 66472 Diesel 91000 Diesel 69700 Diesel 39000 Diesel 39287 Diesel 40569 Diesel 37419 Petrol 34452 Diesel 41567 Hybrid null Hybrid 37778 Diesel 44325 Diesel 45000 Diesel 75000 Diesel 36000 Diesel 64000 Diesel 106000 Diesel 67150 Diesel 43426 Petrol 42622 Diesel 34903 Diesel 35345 Diesel 23146 Petrol 16000 Diesel 44000 Petrol 30000 Diesel 22000 Petrol 19000 Petrol 90000 Diesel 32000 Diesel 90000 Diesel 37000 Diesel 23000 Diesel 78000 Diesel 47000 Diesel 18000 Diesel 64000 Diesel 50000 Diesel 65000 Diesel 38000 Diesel 70000 Diesel 69000 Diesel 41000 Petrol 16250 Petrol 30676 Diesel 40005 Diesel 38000 Hybrid 45000 Diesel 40005 Diesel 120000 Diesel 51936 Diesel 4000 Diesel 8500 Diesel 119168 Petrol 21800 Petrol 19999 Petrol 1600 Petrol 46221 Diesel 45000 Diesel 40005 Diesel 30676 Diesel 52712 Diesel 59858 Diesel 22 Petrol 28234 Petrol 22803 Diesel 17093 Diesel 8150 Diesel 9344 Diesel null Hybrid 36243 Diesel 33000 Diesel 11300 Petrol 32800 Diesel 37600 Diesel 14000 Petrol 24175 Petrol 27968 Diesel 9255 Diesel 26973 Diesel 12491 Diesel 14999 Diesel 37941 Petrol 11301 Diesel 12629 Diesel 20981 Diesel 23760 Diesel 26216 Diesel 80485 Diesel 18739 Diesel 53453 Diesel 23090 Diesel 26100 Diesel 4285 Diesel 28388 Diesel 16902 Diesel 31227 Diesel 24010 Diesel 15838 Diesel 17657 Hybrid 4977 Diesel 24756 Diesel 41416 Diesel 3814 Petrol 17 Petrol 37080 Diesel 73582 Diesel 23772 Diesel 28808 Diesel 10000 Diesel 26413 Diesel 98000 Diesel 39920 Diesel 10455 Petrol 27644 Diesel 63916 Hybrid 15383 Petrol 21929 Diesel 16900 Diesel 34015 Diesel 2000 Petrol 28000 Petrol 22400 Petrol null Petrol 4500 Petrol 836 Petrol 14462 Diesel 23429 Hybrid 20764 Diesel 31706 Diesel 55000 Diesel 10059 Petrol 10357 Diesel 50374 Diesel 42471 Hybrid 41974 Diesel 20618 Diesel 15536 Petrol 16103 Petrol 14576 Petrol 6198 Diesel 6265 Diesel 21786 Diesel 1558 Petrol 2258 Petrol 26866 Diesel 1887 Petrol 9676 Diesel 7894 Diesel 26179 Diesel 21945 Petrol 2195 Diesel 16291 Petrol 38438 Diesel 25107 Hybrid 20491 Hybrid 52792 Diesel 30035 Petrol 3620 Hybrid 19204 Diesel 10 Petrol 46155 Diesel 49494 Diesel 43014 Diesel 29283 Diesel 38830 Diesel 49761 Diesel 31801 Diesel 22349 Petrol 8520 Diesel 27493 Hybrid 15075 Petrol 5116 Diesel 15009 Petrol 6787 Diesel 32409 Diesel 23690 Diesel 101 Diesel 22485 Hybrid 41921 Petrol 35488 Diesel 41647 Diesel 24923 Diesel 11810 Diesel 2000 Diesel 6668 Diesel 19562 Petrol 29201 Diesel 19864 Diesel 22117 Petrol 43331 Diesel 51929 Diesel 42245 Diesel 42062 Petrol 43883 Diesel 36587 Hybrid 42036 Diesel 36538 Diesel 56000 Diesel 11986 Petrol 38783 Diesel 38051 Diesel 10542 Petrol 3033 Petrol 59884 Diesel 1871 Petrol 32594 Diesel 1474 Petrol 15210 Petrol 49783 Diesel 7175 Diesel null Petrol 35576 Diesel 13888 Diesel 4703 Diesel 8143 Diesel 57066 Diesel 25296 Diesel 29907 Diesel 5659 Petrol 23063 Diesel 60948 Diesel 32769 Diesel 20949 Diesel 27890 Diesel 11954 Diesel 42879 Diesel 11470 Diesel 26205 Diesel 61305 Diesel 10561 Diesel 72939 Hybrid 31125 Diesel 37194 Diesel 2240 Petrol 21023 Diesel 35975 Diesel 55226 Diesel 3686 Diesel 33943 Petrol 9699 Petrol 25059 Diesel 101 Diesel 7006 Petrol 38671 Diesel 19877 Hybrid 7937 Petrol 20027 Diesel 19953 Diesel 18551 Diesel 14288 Diesel 10485 Petrol 101 Diesel 27424 Diesel 2985 Diesel 36051 Diesel 3458 Petrol 6406 Diesel 65000 Diesel 22773 Diesel 37000 Petrol 30487 Diesel 29814 Diesel 28050 Diesel 18882 Diesel 27034 Diesel 40620 Diesel 9057 Diesel 7060 Diesel 4615 Diesel 9250 Petrol 1052 Diesel 25149 Petrol 5947 Petrol 2749 Diesel 7144 Diesel 4843 Diesel 44641 Diesel 26682 Diesel 18865 Diesel 21225 Diesel 52519 Diesel 42000 Diesel 3233 Petrol 28535 Diesel 32629 Diesel 30748 Diesel 17600 

> Same for mileage, distribution is **right skewed** for all fuelType features.

In [0]:
%sql
SELECT price, mileage FROM cclass_csv

price mileage 30495 1200 29989 1000 37899 500 30399 5000 29899 4500 30999 null 35999 500 37990 1412 28990 3569 28990 3635 9995 44900 6995 88200 7495 115000 8995 69250 14995 49850 8595 82685 19250 32506 23500 17000 24995 17744 29995 64291 17990 41011 14477 72000 10977 72000 17977 40000 12790 52000 14488 66472 7500 91000 9750 69700 15995 39000 15675 39287 16475 40569 15975 37419 16475 34452 15975 41567 16495 null 17795 37778 17875 44325 14000 45000 16500 75000 21000 36000 15500 64000 7250 106000 13000 67150 11000 43426 13300 42622 16000 34903 12500 35345 14495 23146 24500 16000 23990 44000 17250 30000 14200 22000 22950 19000 10500 90000 15650 32000 14850 90000 22400 37000 19950 23000 9990 78000 11990 47000 20990 18000 14490 64000 10490 50000 15990 65000 17850 38000 16450 70000 11350 69000 12900 41000 26000 16250 22767 30676 12000 40005 18999 38000 14000 45000 12000 40005 4343 120000 16500 51936 27750 4000 28950 8500 3495 119168 59740 21800 44990 19999 54990 1600 15995 46221 14000 45000 12000 40005 22767 30676 12998 52712 17998 59858 28998 22 24998 28234 19498 22803 23498 17093 25498 8150 24998 9344 18498 null 15498 36243 20995 33000 31995 11300 23995 32800 16995 37600 24995 14000 26998 24175 20498 27968 27498 9255 16998 26973 27498 12491 27698 14999 27498 37941 32498 11301 24498 12629 21498 20981 18998 23760 15998 26216 9298 80485 22998 18739 12998 53453 22998 23090 17998 26100 28498 4285 18998 28388 20998 16902 21998 31227 13998 24010 24498 15838 20998 17657 30498 4977 19498 24756 15298 41416 33998 3814 32998 17 16498 37080 15498 73582 15498 23772 18998 28808 24495 10000 18990 26413 10995 98000 15698 39920 22498 10455 19298 27644 16498 63916 19998 15383 20685 21929 20489 16900 13800 34015 31998 2000 29998 28000 24998 22400 32998 null 32998 4500 29998 836 23998 14462 20998 23429 16498 20764 11998 31706 12495 55000 19298 10059 23998 10357 12495 50374 17498 42471 16198 41974 18498 20618 24998 15536 24998 16103 21998 14576 29498 6198 29498 6265 16998 21786 33998 1558 32998 2258 18998 26866 32498 1887 25798 9676 20498 7894 18498 26179 14498 21945 29498 2195 17998 16291 13498 38438 19698 25107 22498 20491 11998 52792 11998 30035 39998 3620 23498 19204 57998 10 16315 46155 13991 49494 16998 43014 16498 29283 16998 38830 13690 49761 14490 31801 17998 22349 29998 8520 18998 27493 18998 15075 28998 5116 24998 15009 27998 6787 16798 32409 19998 23690 29950 101 20790 22485 9490 41921 15690 35488 14990 41647 17290 24923 19490 11810 28990 2000 24990 6668 28990 19562 16298 29201 14698 19864 13998 22117 14698 43331 10498 51929 11998 42245 15698 42062 16998 43883 17998 36587 20998 42036 13691 36538 13491 56000 21998 11986 14500 38783 15990 38051 20490 10542 20890 3033 14490 59884 21990 1871 14750 32594 44990 1474 24690 15210 18270 49783 26625 7175 13361 null 17490 35576 26427 13888 29290 4703 23047 8143 13621 57066 18200 25296 16991 29907 23498 5659 14698 23063 9798 60948 17998 32769 18998 20949 20700 27890 16990 11954 11791 42879 18191 11470 20991 26205 13321 61305 21998 10561 15998 72939 13498 31125 20871 37194 25891 2240 20591 21023 19991 35975 15990 55226 30055 3686 21650 33943 25990 9699 17690 25059 30950 101 24950 7006 14998 38671 19798 19877 24498 7937 19991 20027 21991 19953 22591 18551 18790 14288 25950 10485 29950 101 14990 27424 31950 2985 20990 36051 28990 3458 27000 6406 12590 65000 20885 22773 25885 37000 14965 30487 16595 29814 17940 28050 19940 18882 16498 27034 14500 40620 27791 9057 26499 7060 31990 4615 20999 9250 31999 1052 17495 25149 25495 5947 36995 2749 30999 7144 29999 4843 17880 44641 13660 26682 23920 18865 22960 21225 16590 52519 17990 42000 22270 3233 22680 28535 22440 32629 17820 30748 19290 17600 16390 31360 14450 19598 17250 28192 19690 38087 19250 23523 21980 7716 23990 1647 21980 5572 27990 101 16998 66796 19698 41056 15498 32140 67999 1500 34999 1500 21299 12249 17999 37564 19999 14360 21999 10288 17699 43236 14580 33113 23990 19784 28660 1610 19490 31326 21770 32790 25990 11728 28660 6282 22590 41131 22770 14

> Typical for all used cars listings, price depend strongly by mileage, low mileage the price of car is higher.

In [0]:
%sql
SELECT transmission FROM cclass_csv

transmission Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Automatic Automatic Automatic Automatic Manual Automatic Automatic Manual Manual Manual Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Manual Automatic Automatic Manual Manual Automatic Manual Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Automatic Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Manual Automatic Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Manual Manual Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Manual Automatic Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Automatic Semi-Auto Semi-Auto Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Manual Automatic Manual Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Automatic Automatic Automatic Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Automatic Semi-Auto Automatic Semi-Auto Automatic Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto Semi-Auto S

In [0]:
%sql
SELECT price, engineSize FROM cclass_csv

price engineSize 30495 2.0 29989 1.5 37899 2.0 30399 2.0 29899 2.0 30999 2.0 35999 2.0 37990 3.0 28990 2.0 28990 2.0 9995 1.6 6995 2.1 7495 2.1 8995 2.1 14995 2.1 8595 2.1 19250 2.1 23500 2.1 24995 2.0 29995 4.0 17990 2.0 14477 1.6 10977 2.1 17977 1.6 12790 2.1 14488 2.1 7500 2.1 9750 2.1 15995 1.6 15675 2.1 16475 2.1 15975 2.0 16475 2.1 15975 2.1 16495 2.0 17795 2.1 17875 2.1 14000 2.1 16500 1.6 21000 2.1 15500 2.1 7250 2.1 13000 2.1 11000 1.6 13300 2.1 16000 2.1 12500 2.1 14495 2.0 24500 2.1 23990 6.2 17250 2.1 14200 1.6 22950 2.0 10500 2.1 15650 2.1 14850 2.1 22400 2.1 19950 2.1 9990 2.1 11990 2.1 20990 2.1 14490 2.1 10490 2.1 15990 2.1 17850 2.1 16450 2.1 11350 2.1 12900 1.6 26000 1.5 22767 2.1 12000 2.1 18999 2.0 14000 2.1 12000 2.1 4343 2.1 16500 2.1 27750 2.0 28950 2.0 3495 1.8 59740 4.0 44990 4.0 54990 4.0 15995 2.1 14000 2.1 12000 2.1 22767 2.1 12998 2.1 17998 2.1 28998 1.6 24998 2.0 19498 2.1 23498 2.1 25498 1.6 24998 2.0 18498 2.0 15498 2.1 20995 2.1 31995 3.0 23995 2.1 16995 2.1 24995 2.0 26998 6.2 20498 2.1 27498 2.0 16998 1.6 27498 2.1 27698 2.0 27498 3.0 32498 2.0 24498 2.1 21498 2.1 18998 2.1 15998 2.1 9298 2.1 22998 2.1 12998 2.1 22998 2.1 17998 2.1 28498 2.0 18998 1.6 20998 2.1 21998 2.1 13998 1.6 24498 2.0 20998 2.1 30498 2.0 19498 2.1 15298 2.1 33998 1.5 32998 1.5 16498 2.1 15498 2.1 15498 2.1 18998 2.1 24495 2.0 18990 2.1 10995 2.1 15698 2.1 22498 1.5 19298 2.1 16498 2.0 19998 1.5 20685 2.1 20489 2.1 13800 2.1 31998 2.0 29998 2.0 24998 1.5 32998 2.0 32998 2.0 29998 2.0 23998 2.1 20998 2.0 16498 2.1 11998 2.1 12495 2.1 19298 1.5 23998 2.0 12495 2.1 17498 2.1 16198 2.1 18498 2.1 24998 1.5 24998 1.5 21998 1.5 29498 2.0 29498 2.0 16998 2.1 33998 1.5 32998 1.5 18998 2.1 32498 1.5 25798 2.1 20498 2.1 18498 2.1 14498 2.0 29498 2.0 17998 2.0 13498 2.1 19698 2.0 22498 2.1 11998 1.6 11998 1.6 39998 2.0 23498 2.0 57998 4.0 16315 2.1 13991 2.1 16998 2.1 16498 2.1 16998 2.1 13690 2.1 14490 2.1 17998 2.0 29998 2.0 18998 2.1 18998 2.0 28998 2.0 24998 1.5 27998 2.0 16798 2.1 19998 2.1 29950 2.0 20790 2.1 9490 1.6 15690 2.1 14990 2.1 17290 2.1 19490 2.1 28990 2.0 24990 2.1 28990 6.2 16298 1.6 14698 2.1 13998 2.0 14698 2.1 10498 2.1 11998 2.1 15698 2.0 16998 2.1 17998 2.0 20998 2.1 13691 1.6 13491 2.1 21998 2.0 14500 2.1 15990 2.1 20490 2.0 20890 1.5 14490 2.1 21990 1.5 14750 2.1 44990 3.0 24690 1.5 18270 2.1 26625 2.0 13361 2.0 17490 2.1 26427 2.0 29290 2.0 23047 2.1 13621 1.6 18200 2.1 16991 2.1 23498 2.0 14698 2.1 9798 2.1 17998 2.1 18998 2.1 20700 2.1 16990 2.1 11791 2.1 18191 1.6 20991 2.1 13321 1.6 21998 2.1 15998 2.0 13498 2.1 20871 2.1 25891 2.0 20591 2.1 19991 2.1 15990 2.1 30055 2.0 21650 2.0 25990 2.0 17690 2.1 30950 2.0 24950 1.5 14998 2.1 19798 2.0 24498 1.5 19991 2.1 21991 2.1 22591 2.1 18790 2.1 25950 1.5 29950 2.0 14990 2.1 31950 2.0 20990 2.1 28990 1.5 27000 2.0 12590 2.1 20885 2.1 25885 6.2 14965 2.1 16595 2.1 17940 2.1 19940 2.1 16498 1.6 14500 1.6 27791 2.0 26499 2.0 31990 2.0 20999 1.5 31999 2.0 17495 2.0 25495 1.6 36995 2.0 30999 2.0 29999 2.0 17880 2.1 13660 2.1 23920 2.1 22960 2.1 16590 1.6 17990 2.1 22270 1.6 22680 2.1 22440 2.1 17820 2.1 19290 2.1 16390 2.1 14450 2.0 17250 2.1 19690 2.1 19250 2.1 21980 1.5 23990 1.5 21980 1.5 27990 2.0 16998 2.1 19698 2.1 15498 2.1 67999 4.0 34999 1.5 21299 2.0 17999 2.0 19999 1.5 21999 2.0 17699 2.1 14580 2.1 23990 2.0 28660 2.0 19490 2.1 21770 2.1 25990 2.0 28660 2.0 22590 2.1 22770 2.1 16480 2.1 27698 2.0 16499 2.0 19895 1.5 27595 2.0 25795 2.0 24499 1.5 30495 2.0 29999 2.0 27495 2.0 21795 1.6 9881 1.6 13498 2.0 24450 2.0 17450 2.1 14450 2.1 24990 2.1 16440 2.1 9891 2.1 16998 2.0 16710 2.1 22495 2.1 21565 2.0 21697 2.0 20990 2.1 14770 2.0 13448 2.1 15998 2.1 27498 2.0 19498 2.0 22498 2.1 20105 2.1 14220 2.1 22260 2.1 16990 2.1 41880 3.0 30990 2.0 28880 1.5 28990 2.0 18700 2.1 17440 2.1 17770 2.1 18433 2.1 21440 2.1 22170 2.1 21115 2.1 21497 1.6 28400 2.0 22765 2.1 22880 2.1 22455 1.5 17330 2.1 18795 2.1 15220 2.1 21440 2.1 20200 2.1 19660 1.6 22990 2.1 22770 2.1

In [0]:
%sql
SELECT mileage, transmission FROM cclass_csv

mileage transmission 1200 Automatic 1000 Automatic 500 Automatic 5000 Automatic 4500 Automatic null Automatic 500 Automatic 1412 Automatic 3569 Automatic 3635 Automatic 44900 Automatic 88200 Automatic 115000 Automatic 69250 Automatic 49850 Automatic 82685 Automatic 32506 Semi-Auto 17000 Semi-Auto 17744 Semi-Auto 64291 Semi-Auto 41011 Automatic 72000 Automatic 72000 Automatic 40000 Automatic 52000 Automatic 66472 Automatic 91000 Manual 69700 Automatic 39000 Automatic 39287 Manual 40569 Manual 37419 Manual 34452 Automatic 41567 Automatic null Automatic 37778 Automatic 44325 Automatic 45000 Automatic 75000 Automatic 36000 Automatic 64000 Automatic 106000 Manual 67150 Automatic 43426 Automatic 42622 Manual 34903 Manual 35345 Automatic 23146 Manual 16000 Automatic 44000 Automatic 30000 Automatic 22000 Automatic 19000 Automatic 90000 Automatic 32000 Automatic 90000 Automatic 37000 Automatic 23000 Automatic 78000 Automatic 47000 Automatic 18000 Automatic 64000 Automatic 50000 Automatic 65000 Automatic 38000 Automatic 70000 Automatic 69000 Automatic 41000 Automatic 16250 Automatic 30676 Automatic 40005 Automatic 38000 Automatic 45000 Automatic 40005 Automatic 120000 Automatic 51936 Automatic 4000 Automatic 8500 Automatic 119168 Automatic 21800 Automatic 19999 Automatic 1600 Automatic 46221 Automatic 45000 Automatic 40005 Automatic 30676 Automatic 52712 Semi-Auto 59858 Semi-Auto 22 Manual 28234 Semi-Auto 22803 Semi-Auto 17093 Semi-Auto 8150 Semi-Auto 9344 Semi-Auto null Semi-Auto 36243 Semi-Auto 33000 Semi-Auto 11300 Semi-Auto 32800 Semi-Auto 37600 Automatic 14000 Automatic 24175 Semi-Auto 27968 Semi-Auto 9255 Semi-Auto 26973 Semi-Auto 12491 Semi-Auto 14999 Semi-Auto 37941 Semi-Auto 11301 Semi-Auto 12629 Semi-Auto 20981 Semi-Auto 23760 Semi-Auto 26216 Semi-Auto 80485 Semi-Auto 18739 Semi-Auto 53453 Manual 23090 Semi-Auto 26100 Manual 4285 Semi-Auto 28388 Semi-Auto 16902 Semi-Auto 31227 Semi-Auto 24010 Manual 15838 Semi-Auto 17657 Semi-Auto 4977 Semi-Auto 24756 Semi-Auto 41416 Semi-Auto 3814 Semi-Auto 17 Semi-Auto 37080 Semi-Auto 73582 Semi-Auto 23772 Semi-Auto 28808 Semi-Auto 10000 Semi-Auto 26413 Semi-Auto 98000 Manual 39920 Semi-Auto 10455 Semi-Auto 27644 Semi-Auto 63916 Semi-Auto 15383 Semi-Auto 21929 Semi-Auto 16900 Semi-Auto 34015 Manual 2000 Semi-Auto 28000 Semi-Auto 22400 Semi-Auto null Semi-Auto 4500 Semi-Auto 836 Semi-Auto 14462 Semi-Auto 23429 Semi-Auto 20764 Automatic 31706 Semi-Auto 55000 Automatic 10059 Automatic 10357 Semi-Auto 50374 Semi-Auto 42471 Semi-Auto 41974 Manual 20618 Semi-Auto 15536 Semi-Auto 16103 Semi-Auto 14576 Semi-Auto 6198 Semi-Auto 6265 Semi-Auto 21786 Semi-Auto 1558 Semi-Auto 2258 Semi-Auto 26866 Semi-Auto 1887 Semi-Auto 9676 Semi-Auto 7894 Semi-Auto 26179 Semi-Auto 21945 Manual 2195 Semi-Auto 16291 Semi-Auto 38438 Manual 25107 Automatic 20491 Semi-Auto 52792 Manual 30035 Semi-Auto 3620 Semi-Auto 19204 Semi-Auto 10 Semi-Auto 46155 Semi-Auto 49494 Semi-Auto 43014 Semi-Auto 29283 Semi-Auto 38830 Semi-Auto 49761 Manual 31801 Semi-Auto 22349 Semi-Auto 8520 Semi-Auto 27493 Semi-Auto 15075 Semi-Auto 5116 Semi-Auto 15009 Semi-Auto 6787 Semi-Auto 32409 Semi-Auto 23690 Semi-Auto 101 Semi-Auto 22485 Semi-Auto 41921 Semi-Auto 35488 Semi-Auto 41647 Manual 24923 Semi-Auto 11810 Semi-Auto 2000 Semi-Auto 6668 Automatic 19562 Semi-Auto 29201 Semi-Auto 19864 Manual 22117 Manual 43331 Semi-Auto 51929 Semi-Auto 42245 Semi-Auto 42062 Automatic 43883 Semi-Auto 36587 Semi-Auto 42036 Semi-Auto 36538 Semi-Auto 56000 Manual 11986 Semi-Auto 38783 Semi-Auto 38051 Semi-Auto 10542 Semi-Auto 3033 Semi-Auto 59884 Semi-Auto 1871 Semi-Auto 32594 Semi-Auto 1474 Automatic 15210 Semi-Auto 49783 Semi-Auto 7175 Automatic null Semi-Auto 35576 Semi-Auto 13888 Automatic 4703 Semi-Auto 8143 Semi-Auto 57066 Semi-Auto 25296 Semi-Auto 29907 Semi-Auto 5659 Semi-Auto 23063 Semi-Auto 60948 Semi-Auto 32769 Automatic 20949 Automatic 27890 Semi-Auto 11954 Semi-Auto 42879 Semi-Auto 11470 Semi-Auto 26205 Semi-Auto 61305 Semi-Auto 10561 Semi-Auto 72939 Sem

> Graphs by boxplot shows outliers per each transmission type above 60k miles.

In [0]:
%sql
SELECT price, fuelType FROM cclass_csv

price fuelType 30495 Diesel 29989 Petrol 37899 Diesel 30399 Diesel 29899 Diesel 30999 Diesel 35999 Diesel 37990 Petrol 28990 Diesel 28990 Diesel 9995 Petrol 6995 Diesel 7495 Diesel 8995 Diesel 14995 Diesel 8595 Diesel 19250 Diesel 23500 Diesel 24995 Diesel 29995 Petrol 17990 Petrol 14477 Diesel 10977 Diesel 17977 Diesel 12790 Diesel 14488 Diesel 7500 Diesel 9750 Diesel 15995 Diesel 15675 Diesel 16475 Diesel 15975 Petrol 16475 Diesel 15975 Hybrid 16495 Hybrid 17795 Diesel 17875 Diesel 14000 Diesel 16500 Diesel 21000 Diesel 15500 Diesel 7250 Diesel 13000 Diesel 11000 Petrol 13300 Diesel 16000 Diesel 12500 Diesel 14495 Petrol 24500 Diesel 23990 Petrol 17250 Diesel 14200 Petrol 22950 Petrol 10500 Diesel 15650 Diesel 14850 Diesel 22400 Diesel 19950 Diesel 9990 Diesel 11990 Diesel 20990 Diesel 14490 Diesel 10490 Diesel 15990 Diesel 17850 Diesel 16450 Diesel 11350 Diesel 12900 Petrol 26000 Petrol 22767 Diesel 12000 Diesel 18999 Hybrid 14000 Diesel 12000 Diesel 4343 Diesel 16500 Diesel 27750 Diesel 28950 Diesel 3495 Petrol 59740 Petrol 44990 Petrol 54990 Petrol 15995 Diesel 14000 Diesel 12000 Diesel 22767 Diesel 12998 Diesel 17998 Diesel 28998 Petrol 24998 Petrol 19498 Diesel 23498 Diesel 25498 Diesel 24998 Diesel 18498 Hybrid 15498 Diesel 20995 Diesel 31995 Petrol 23995 Diesel 16995 Diesel 24995 Petrol 26998 Petrol 20498 Diesel 27498 Diesel 16998 Diesel 27498 Diesel 27698 Diesel 27498 Petrol 32498 Diesel 24498 Diesel 21498 Diesel 18998 Diesel 15998 Diesel 9298 Diesel 22998 Diesel 12998 Diesel 22998 Diesel 17998 Diesel 28498 Diesel 18998 Diesel 20998 Diesel 21998 Diesel 13998 Diesel 24498 Diesel 20998 Hybrid 30498 Diesel 19498 Diesel 15298 Diesel 33998 Petrol 32998 Petrol 16498 Diesel 15498 Diesel 15498 Diesel 18998 Diesel 24495 Diesel 18990 Diesel 10995 Diesel 15698 Diesel 22498 Petrol 19298 Diesel 16498 Hybrid 19998 Petrol 20685 Diesel 20489 Diesel 13800 Diesel 31998 Petrol 29998 Petrol 24998 Petrol 32998 Petrol 32998 Petrol 29998 Petrol 23998 Diesel 20998 Hybrid 16498 Diesel 11998 Diesel 12495 Diesel 19298 Petrol 23998 Diesel 12495 Diesel 17498 Hybrid 16198 Diesel 18498 Diesel 24998 Petrol 24998 Petrol 21998 Petrol 29498 Diesel 29498 Diesel 16998 Diesel 33998 Petrol 32998 Petrol 18998 Diesel 32498 Petrol 25798 Diesel 20498 Diesel 18498 Diesel 14498 Petrol 29498 Diesel 17998 Petrol 13498 Diesel 19698 Hybrid 22498 Hybrid 11998 Diesel 11998 Petrol 39998 Hybrid 23498 Diesel 57998 Petrol 16315 Diesel 13991 Diesel 16998 Diesel 16498 Diesel 16998 Diesel 13690 Diesel 14490 Diesel 17998 Petrol 29998 Diesel 18998 Hybrid 18998 Petrol 28998 Diesel 24998 Petrol 27998 Diesel 16798 Diesel 19998 Diesel 29950 Diesel 20790 Hybrid 9490 Petrol 15690 Diesel 14990 Diesel 17290 Diesel 19490 Diesel 28990 Diesel 24990 Diesel 28990 Petrol 16298 Diesel 14698 Diesel 13998 Petrol 14698 Diesel 10498 Diesel 11998 Diesel 15698 Petrol 16998 Diesel 17998 Hybrid 20998 Diesel 13691 Diesel 13491 Diesel 21998 Petrol 14500 Diesel 15990 Diesel 20490 Petrol 20890 Petrol 14490 Diesel 21990 Petrol 14750 Diesel 44990 Petrol 24690 Petrol 18270 Diesel 26625 Diesel 13361 Petrol 17490 Diesel 26427 Diesel 29290 Diesel 23047 Diesel 13621 Diesel 18200 Diesel 16991 Diesel 23498 Petrol 14698 Diesel 9798 Diesel 17998 Diesel 18998 Diesel 20700 Diesel 16990 Diesel 11791 Diesel 18191 Diesel 20991 Diesel 13321 Diesel 21998 Diesel 15998 Hybrid 13498 Diesel 20871 Diesel 25891 Petrol 20591 Diesel 19991 Diesel 15990 Diesel 30055 Diesel 21650 Petrol 25990 Petrol 17690 Diesel 30950 Diesel 24950 Petrol 14998 Diesel 19798 Hybrid 24498 Petrol 19991 Diesel 21991 Diesel 22591 Diesel 18790 Diesel 25950 Petrol 29950 Diesel 14990 Diesel 31950 Diesel 20990 Diesel 28990 Petrol 27000 Diesel 12590 Diesel 20885 Diesel 25885 Petrol 14965 Diesel 16595 Diesel 17940 Diesel 19940 Diesel 16498 Diesel 14500 Diesel 27791 Diesel 26499 Diesel 31990 Diesel 20999 Petrol 31999 Diesel 17495 Petrol 25495 Petrol 36995 Diesel 30999 Diesel 29999 Diesel 17880 Diesel 13660 Diesel 23920 Diesel 22960 Diesel 16590 Diesel 17990 Di

> As mileage feature above 40k, there are outliers by price.

In [0]:
%sql
SELECT * FROM cclass_csv

model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 29989 Automatic 1000 Petrol 1.5 C Class 2020-01-01 37899 Automatic 500 Diesel 2.0 C Class 2019-01-01 30399 Automatic 5000 Diesel 2.0 C Class 2019-01-01 29899 Automatic 4500 Diesel 2.0 C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2020-01-01 35999 Automatic 500 Diesel 2.0 C Class 2019-01-01 37990 Automatic 1412 Petrol 3.0 C Class 2019-01-01 28990 Automatic 3569 Diesel 2.0 C Class 2019-01-01 28990 Automatic 3635 Diesel 2.0 C Class 2013-01-01 9995 Automatic 44900 Petrol 1.6 C Class 2012-01-01 6995 Automatic 88200 Diesel 2.1 C Class 2012-01-01 7495 Automatic 115000 Diesel 2.1 C Class 2011-01-01 8995 Automatic 69250 Diesel 2.1 C Class 2015-01-01 14995 Automatic 49850 Diesel 2.1 C Class 2013-01-01 8595 Automatic 82685 Diesel 2.1 C Class 2016-01-01 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017-01-01 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018-01-01 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016-01-01 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014-01-01 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013-01-01 10977 Automatic 72000 Diesel 2.1 C Class 2016-01-01 17977 Automatic 40000 Diesel 1.6 C Class 2015-01-01 12790 Automatic 52000 Diesel 2.1 C Class 2017-01-01 14488 Automatic 66472 Diesel 2.1 C Class 2011-01-01 7500 Manual 91000 Diesel 2.1 C Class 2013-01-01 9750 Automatic 69700 Diesel 2.1 C Class 2015-01-01 15995 Automatic 39000 Diesel 1.6 C Class 2016-01-01 15675 Manual 39287 Diesel 2.1 C Class 2016-01-01 16475 Manual 40569 Diesel 2.1 C Class 2016-01-01 15975 Manual 37419 Petrol 2.0 C Class 2016-01-01 16475 Automatic 34452 Diesel 2.1 C Class 2015-01-01 15975 Automatic 41567 Hybrid 2.1 C Class 2015-01-01 16495 Automatic null Hybrid 2.0 C Class 2017-01-01 17795 Automatic 37778 Diesel 2.1 C Class 2017-01-01 17875 Automatic 44325 Diesel 2.1 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2017-01-01 16500 Automatic 75000 Diesel 1.6 C Class 2016-01-01 21000 Automatic 36000 Diesel 2.1 C Class 2015-01-01 15500 Automatic 64000 Diesel 2.1 C Class 2013-01-01 7250 Manual 106000 Diesel 2.1 C Class 2015-01-01 13000 Automatic 67150 Diesel 2.1 C Class 2012-01-01 11000 Automatic 43426 Petrol 1.6 C Class 2016-01-01 13300 Manual 42622 Diesel 2.1 C Class 2016-01-01 16000 Manual 34903 Diesel 2.1 C Class 2014-01-01 12500 Automatic 35345 Diesel 2.1 C Class 2015-01-01 14495 Manual 23146 Petrol 2.0 C Class 2017-01-01 24500 Automatic 16000 Diesel 2.1 C Class 2013-01-01 23990 Automatic 44000 Petrol 6.2 C Class 2015-01-01 17250 Automatic 30000 Diesel 2.1 C Class 2015-01-01 14200 Automatic 22000 Petrol 1.6 C Class 2017-01-01 22950 Automatic 19000 Petrol 2.0 C Class 2013-01-01 10500 Automatic 90000 Diesel 2.1 C Class 2014-01-01 15650 Automatic 32000 Diesel 2.1 C Class 2015-01-01 14850 Automatic 90000 Diesel 2.1 C Class 2017-01-01 22400 Automatic 37000 Diesel 2.1 C Class 2017-01-01 19950 Automatic 23000 Diesel 2.1 C Class 2014-01-01 9990 Automatic 78000 Diesel 2.1 C Class 2015-01-01 11990 Automatic 47000 Diesel 2.1 C Class 2017-01-01 20990 Automatic 18000 Diesel 2.1 C Class 2015-01-01 14490 Automatic 64000 Diesel 2.1 C Class 2013-01-01 10490 Automatic 50000 Diesel 2.1 C Class 2016-01-01 15990 Automatic 65000 Diesel 2.1 C Class 2016-01-01 17850 Automatic 38000 Diesel 2.1 C Class 2016-01-01 16450 Automatic 70000 Diesel 2.1 C Class 2013-01-01 11350 Automatic 69000 Diesel 2.1 C Class 2014-01-01 12900 Automatic 41000 Petrol 1.6 C Class 2018-01-01 26000 Automatic 16250 Petrol 1.5 C Class 2017-01-01 22767 Automatic 30676 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2017-01-01 18999 Automatic 38000 Hybrid 2.0 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2009-01-01 4343 Automatic 120000 Diesel 2.1 C Class 2016-01-01 16500 Automatic 51936 Diesel 2.1 C Class 2019-01-01 27750 Automatic 4000 

In [0]:
%sql
SELECT * FROM cclass_csv 

model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 29989 Automatic 1000 Petrol 1.5 C Class 2020-01-01 37899 Automatic 500 Diesel 2.0 C Class 2019-01-01 30399 Automatic 5000 Diesel 2.0 C Class 2019-01-01 29899 Automatic 4500 Diesel 2.0 C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2020-01-01 35999 Automatic 500 Diesel 2.0 C Class 2019-01-01 37990 Automatic 1412 Petrol 3.0 C Class 2019-01-01 28990 Automatic 3569 Diesel 2.0 C Class 2019-01-01 28990 Automatic 3635 Diesel 2.0 C Class 2013-01-01 9995 Automatic 44900 Petrol 1.6 C Class 2012-01-01 6995 Automatic 88200 Diesel 2.1 C Class 2012-01-01 7495 Automatic 115000 Diesel 2.1 C Class 2011-01-01 8995 Automatic 69250 Diesel 2.1 C Class 2015-01-01 14995 Automatic 49850 Diesel 2.1 C Class 2013-01-01 8595 Automatic 82685 Diesel 2.1 C Class 2016-01-01 19250 Semi-Auto 32506 Diesel 2.1 C Class 2017-01-01 23500 Semi-Auto 17000 Diesel 2.1 C Class 2018-01-01 24995 Semi-Auto 17744 Diesel 2.0 C Class 2016-01-01 29995 Semi-Auto 64291 Petrol 4.0 C Class 2014-01-01 17990 Automatic 41011 Petrol 2.0 C Class null 14477 Automatic 72000 Diesel 1.6 C Class 2013-01-01 10977 Automatic 72000 Diesel 2.1 C Class 2016-01-01 17977 Automatic 40000 Diesel 1.6 C Class 2015-01-01 12790 Automatic 52000 Diesel 2.1 C Class 2017-01-01 14488 Automatic 66472 Diesel 2.1 C Class 2011-01-01 7500 Manual 91000 Diesel 2.1 C Class 2013-01-01 9750 Automatic 69700 Diesel 2.1 C Class 2015-01-01 15995 Automatic 39000 Diesel 1.6 C Class 2016-01-01 15675 Manual 39287 Diesel 2.1 C Class 2016-01-01 16475 Manual 40569 Diesel 2.1 C Class 2016-01-01 15975 Manual 37419 Petrol 2.0 C Class 2016-01-01 16475 Automatic 34452 Diesel 2.1 C Class 2015-01-01 15975 Automatic 41567 Hybrid 2.1 C Class 2015-01-01 16495 Automatic null Hybrid 2.0 C Class 2017-01-01 17795 Automatic 37778 Diesel 2.1 C Class 2017-01-01 17875 Automatic 44325 Diesel 2.1 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2017-01-01 16500 Automatic 75000 Diesel 1.6 C Class 2016-01-01 21000 Automatic 36000 Diesel 2.1 C Class 2015-01-01 15500 Automatic 64000 Diesel 2.1 C Class 2013-01-01 7250 Manual 106000 Diesel 2.1 C Class 2015-01-01 13000 Automatic 67150 Diesel 2.1 C Class 2012-01-01 11000 Automatic 43426 Petrol 1.6 C Class 2016-01-01 13300 Manual 42622 Diesel 2.1 C Class 2016-01-01 16000 Manual 34903 Diesel 2.1 C Class 2014-01-01 12500 Automatic 35345 Diesel 2.1 C Class 2015-01-01 14495 Manual 23146 Petrol 2.0 C Class 2017-01-01 24500 Automatic 16000 Diesel 2.1 C Class 2013-01-01 23990 Automatic 44000 Petrol 6.2 C Class 2015-01-01 17250 Automatic 30000 Diesel 2.1 C Class 2015-01-01 14200 Automatic 22000 Petrol 1.6 C Class 2017-01-01 22950 Automatic 19000 Petrol 2.0 C Class 2013-01-01 10500 Automatic 90000 Diesel 2.1 C Class 2014-01-01 15650 Automatic 32000 Diesel 2.1 C Class 2015-01-01 14850 Automatic 90000 Diesel 2.1 C Class 2017-01-01 22400 Automatic 37000 Diesel 2.1 C Class 2017-01-01 19950 Automatic 23000 Diesel 2.1 C Class 2014-01-01 9990 Automatic 78000 Diesel 2.1 C Class 2015-01-01 11990 Automatic 47000 Diesel 2.1 C Class 2017-01-01 20990 Automatic 18000 Diesel 2.1 C Class 2015-01-01 14490 Automatic 64000 Diesel 2.1 C Class 2013-01-01 10490 Automatic 50000 Diesel 2.1 C Class 2016-01-01 15990 Automatic 65000 Diesel 2.1 C Class 2016-01-01 17850 Automatic 38000 Diesel 2.1 C Class 2016-01-01 16450 Automatic 70000 Diesel 2.1 C Class 2013-01-01 11350 Automatic 69000 Diesel 2.1 C Class 2014-01-01 12900 Automatic 41000 Petrol 1.6 C Class 2018-01-01 26000 Automatic 16250 Petrol 1.5 C Class 2017-01-01 22767 Automatic 30676 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2017-01-01 18999 Automatic 38000 Hybrid 2.0 C Class 2016-01-01 14000 Automatic 45000 Diesel 2.1 C Class 2015-01-01 12000 Automatic 40005 Diesel 2.1 C Class 2009-01-01 4343 Automatic 120000 Diesel 2.1 C Class 2016-01-01 16500 Automatic 51936 Diesel 2.1 C Class 2019-01-01 27750 Automatic 4000 

> Car used price depending by mileage and age, when both are low, price is higher.

In [0]:
%sql
SELECT date_format(current_date() , 'y')

date_format(current_date(), y) 2022

In [0]:
%sql
SELECT price,
  ROUND(datediff(date_format(current_date(), 'y'), date_format(year,'y'))/366.5,0) as carAge
FROM cclass_csv

price carAge 30495 2 29989 2 37899 2 30399 3 29899 3 30999 2 35999 2 37990 3 28990 3 28990 3 9995 9 6995 10 7495 10 8995 11 14995 7 8595 9 19250 6 23500 5 24995 4 29995 6 17990 8 14477 null 10977 9 17977 6 12790 7 14488 5 7500 11 9750 9 15995 7 15675 6 16475 6 15975 6 16475 6 15975 7 16495 7 17795 5 17875 5 14000 6 16500 5 21000 6 15500 7 7250 9 13000 7 11000 10 13300 6 16000 6 12500 8 14495 7 24500 5 23990 9 17250 7 14200 7 22950 5 10500 9 15650 8 14850 7 22400 5 19950 5 9990 8 11990 7 20990 5 14490 7 10490 9 15990 6 17850 6 16450 6 11350 9 12900 8 26000 4 22767 5 12000 7 18999 5 14000 6 12000 7 4343 13 16500 6 27750 3 28950 3 3495 15 59740 5 44990 3 54990 3 15995 7 14000 6 12000 7 22767 5 12998 8 17998 5 28998 3 24998 5 19498 6 23498 4 25498 3 24998 3 18498 6 15498 8 20995 5 31995 4 23995 5 16995 8 24995 5 26998 8 20498 6 27498 4 16998 6 27498 4 27698 3 27498 6 32498 3 24498 3 21498 5 18998 5 15998 8 9298 9 22998 4 12998 6 22998 6 17998 6 28498 3 18998 6 20998 5 21998 6 13998 7 24498 3 20998 5 30498 3 19498 6 15298 8 33998 3 32998 3 16498 6 15498 6 15498 8 18998 5 24495 3 18990 6 10995 8 15698 8 22498 3 19298 6 16498 5 19998 4 20685 4 20489 4 13800 5 31998 3 29998 3 24998 3 32998 3 32998 3 29998 3 23998 5 20998 6 16498 5 11998 9 12495 7 19298 3 23998 3 12495 7 17498 7 16198 6 18498 5 24998 3 24998 3 21998 3 29498 4 29498 4 16998 6 33998 3 32998 3 18998 5 32498 3 25798 5 20498 4 18498 5 14498 6 29498 3 17998 5 13498 7 19698 6 22498 5 11998 7 11998 null 39998 2 23498 3 57998 3 16315 7 13991 6 16998 6 16498 5 16998 6 13690 6 14490 8 17998 5 29998 3 18998 5 18998 4 28998 3 24998 3 27998 4 16798 7 19998 5 29950 2 20790 5 9490 9 15690 5 14990 5 17290 7 19490 4 28990 3 24990 5 28990 8 16298 6 14698 6 13998 6 14698 7 10498 8 11998 7 15698 6 16998 7 17998 6 20998 5 13691 6 13491 8 21998 5 14500 7 15990 5 20490 6 20890 4 14490 8 21990 4 14750 7 44990 3 24690 3 18270 5 26625 4 13361 7 17490 5 26427 3 29290 3 23047 5 13621 7 18200 5 16991 5 23498 4 14698 7 9798 9 17998 6 18998 5 20700 6 16990 7 11791 8 18191 7 20991 6 13321 5 21998 6 15998 5 13498 7 20871 5 25891 4 20591 5 19991 6 15990 7 30055 3 21650 5 25990 4 17690 5 30950 3 24950 3 14998 7 19798 5 24498 3 19991 4 21991 5 22591 4 18790 3 25950 3 29950 2 14990 5 31950 2 20990 6 28990 2 27000 3 12590 7 20885 4 25885 8 14965 6 16595 5 17940 4 19940 4 16498 6 14500 6 27791 3 26499 3 31990 3 20999 4 31999 3 17495 5 25495 3 36995 3 30999 3 29999 3 17880 6 13660 6 23920 6 22960 5 16590 5 17990 5 22270 3 22680 6 22440 5 17820 6 19290 6 16390 6 14450 6 17250 6 19690 6 19250 5 21980 3 23990 3 21980 3 27990 3 16998 7 19698 6 15498 7 67999 2 34999 2 21299 6 17999 5 19999 3 21999 5 17699 6 14580 6 23990 5 28660 3 19490 5 21770 5 25990 3 28660 3 22590 5 22770 5 16480 7 27698 4 16499 5 19895 3 27595 3 25795 3 24499 3 30495 3 29999 3 27495 4 21795 3 9881 9 13498 7 24450 3 17450 7 14450 6 24990 5 16440 7 9891 8 16998 7 16710 7 22495 5 21565 5 21697 5 20990 6 14770 6 13448 7 15998 8 27498 4 19498 6 22498 5 20105 6 14220 6 22260 5 16990 8 41880 3 30990 3 28880 3 28990 3 18700 6 17440 5 17770 6 18433 5 21440 6 22170 5 21115 5 21497 3 28400 3 22765 6 22880 4 22455 3 17330 5 18795 5 15220 6 21440 5 20200 5 19660 4 22990 5 22770 4 19775 4 24660 5 33495 5 26890 3 26885 3 34790 3 18495 5 15699 6 18499 5 24799 4 18995 5 15280 8 14795 8 19840 4 13791 8 24360 5 16990 6 17990 6 14990 8 15990 6 16998 5 21495 6 20995 6 20495 4 24736 3 21595 5 21295 5 19995 6 23495 5 15490 6 17000 5 16402 5 20682 3 29232 4 14698 7 14288 8 18265 4 15971 8 19465 5 16210 5 16990 6 20990 5 17699 6 18495 5 28995 3 14495 6 13950 8 12750 8 19495 6 18995 6 19950 4 19498 5 16360 8 16767 5 10190 8 19298 4 15998 7 15998 6 24698 5 14398 6 15998 8 19950 4 11149 8 17990 4 16490 5 18450 5 11850 7 16350 5 10477 7 18998 4 15870 6 16700 6 19502 5 28332 3 17992 6 49280 3 28810 3 30000 3 17901 7 19971 5 26281 4 17791 5 16998 6 14498 8 16998 6 16990 7 19798 6 16998 6 16750 5 12880 7 18660 5 17990 6 25882 4 29046 3 26256 3 20498 5 11298 9 16500 5 16990 

> Confirm when lower the car age is, the price is higher.

## 6. Data Wrangling

### 6.1 Missing Values

Many real-world datasets may contain missing values for various reasons. They are often encoded as NaNs, blanks or any other placeholders. Training a model with a dataset that has a lot of missing values can drastically impact the machine learning model’s quality. Some algorithms such as scikit-learn estimators assume that all values are numerical and have and hold meaningful value.

One way to handle this problem is to get rid of the observations that have missing data. However, you will risk losing data points with valuable information. A better strategy would be to impute the missing values. In other words, we need to infer those missing values from the existing part of the data. 

we will focus on 6 popular ways for data imputation for cross-sectional datasets:

* Do Nothing
* Imputation Using (Mean/Median) Values
* Imputation Using (Most Frequent) or (Zero/Constant) Values
* Imputation Using k-NN
* Imputation Using Multivariate Imputation by Chained Equation (MICE)
* Imputation Using Deep Learning (Datawig)

We can check all of them and evaluate model combining each step described before.

#### a. Do Nothing

That’s an easy one. You just let the algorithm handle the missing data. Some algorithms can factor in the missing values and learn the best imputation values for the missing data based on the training loss reduction (ie. XGBoost). Some others have the option to just ignore them (ie. LightGBM — use_missing=false). However, other algorithms will panic and throw an error complaining about the missing values (ie. Scikit learn — LinearRegression). In that case, you will need to handle the missing data and clean it before feeding it to the algorithm.

#### b. Imputation Using (Mean/Median) Values:

This works by calculating the mean/median of the non-missing values in a column and then replacing the missing values within each column separately and independently from the others. It can only be used with numeric data.

*Pros*:
Easy and fast.
Works well with small numerical datasets.

*Cons*:
Doesn’t factor the correlations between features. It only works on the column level.
Will give poor results on encoded categorical features (do NOT use it on categorical features).
Not very accurate.
Doesn’t account for the uncertainty in the imputations

#### c. Imputation Using (Most Frequent) or (Zero/Constant) Values

Most Frequent is another statistical strategy to impute missing values and YES!! It works with categorical features (strings or numerical representations) by replacing missing data with the most frequent values within each column.

*Pros*:
Works well with categorical features.

*Cons*:
It also doesn’t factor the correlations between features.
It can introduce bias in the data.

#### d. Imputation Using k-NN:

The k nearest neighbours is an algorithm that is used for simple classification. The algorithm uses ‘feature similarity’ to predict the values of any new data points. This means that the new point is assigned a value based on how closely it resembles the points in the training set. This can be very useful in making predictions about the missing values by finding the k’s closest neighbours to the observation with missing data and then imputing them based on the non-missing values in the neighbourhood. Let’s see some example code using Impyute library which provides a simple and easy way to use KNN for imputation:

*Pros*:
Can be much more accurate than the mean, median or most frequent imputation methods (It depends on the dataset).

*Cons*:
Computationally expensive. KNN works by storing the whole training dataset in memory.
K-NN is quite sensitive to outliers in the data (unlike SVM)

#### e. Imputation Using Multivariate Imputation by Chained Equation (MICE)

This type of imputation works by filling the missing data multiple times. Multiple Imputations (MIs) are much better than a single imputation as it measures the uncertainty of the missing values in a better way. The chained equations approach is also very flexible and can handle different variables of different data types (ie., continuous or binary) as well as complexities such as bounds or survey skip patterns. For more information on the algorithm mechanics, you can refer to the Research Paper

#### f. Imputation Using Deep Learning (Datawig):

This method works very well with categorical and non-numerical features. It is a library that learns Machine Learning models using Deep Neural Networks to impute missing values in a dataframe. It also supports both CPU and GPU for training.

*Pros*:
Quite accurate compared to other methods.
It has some functions that can handle categorical data (Feature Encoder).
It supports CPUs and GPUs.

*Cons*:
Single Column imputation.
Can be quite slow with large datasets.
You have to specify the columns that contain information about the target column that will be imputed.


#### Other Imputation Methods:

##### Stochastic regression imputation:
It is quite similar to regression imputation which tries to predict the missing values by regressing it from other related variables in the same dataset plus some random residual value.

##### Extrapolation and Interpolation:
It tries to estimate values from other observations within the range of a discrete set of known data points.

##### Hot-Deck imputation:
Works by randomly choosing the missing value from a set of related and similar variables.

In [0]:
%python
# Check null and NA values for both dataset by frequency first and absolute values after
# Create a function to detect and shows missing values
def missing_zero_values_table(df):
        zero_val = (df == 0.00).astype(int).sum(axis=0)
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mz_table = pd.concat([zero_val, mis_val, mis_val_percent], axis=1)
        mz_table = mz_table.rename(
        columns = {0 : 'Zero Values', 1 : 'Missing Values', 2 : '% of Total Values'})
        mz_table['Total Zero Missing Values'] = mz_table['Zero Values'] + mz_table['Missing Values']
        mz_table['% Total Zero Missing Values'] = 100 * mz_table['Total Zero Missing Values'] / len(df)
        mz_table['Data Type'] = df.dtypes
        mz_table = mz_table[
            mz_table.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns and " + str(df.shape[0]) + " Rows.\n"      
            "There are " + str(mz_table.shape[0]) +
              " columns that have missing values.")
        return mz_table


In [0]:
# Check null values
import pandas as pd
dataset = df.toPandas()
missing_zero_values_table(dataset)

Your selected dataframe has 7 columns and 3899 Rows.
There are 2 columns that have missing values.


Zero Values 
 Missing Values 
 % of Total Values 
 Total Zero Missing Values 
 % Total Zero Missing Values 
 Data Type 
 
 
 
 
 mileage 
 0 
 7 
 0.2 
 7 
 0.2 
 float64 
 
 
 year 
 0 
 5 
 0.1 
 5 
 0.1 
 object

> It seems dataset has only 12 missing values, 7 in `mileage` and 5 in `year` feature. We analyze them to take a decision to impute or remove them.

In [0]:
%sql
/*# Check year null values on cross table */
SELECT 
  YEAR(year) as anyo
  ,COUNT(*) as total_values
FROM cclass_csv
GROUP BY 1
ORDER BY 2 DESC;

anyo total_values 2019 1589 2017 665 2016 544 2018 361 2015 321 2014 136 2020 125 2013 63 2012 26 2011 13 2010 10 2008 8 2007 7 2009 7 null 5 2005 5 2002 4 2004 3 2006 3 2003 1 1991 1 1998 1 1995 1

In [0]:
%sql
/* # Check null values and all variables */
SELECT *
FROM cclass_csv
WHERE year is NULL

model year price transmission mileage fuelType engineSize C Class null 14477 Automatic 72000 Diesel 1.6 C Class null 11998 Semi-Auto 30035 Petrol 1.6 C Class null 37995 Semi-Auto 5000 Petrol 1.5 C Class null 28180 Automatic 11487 Diesel 2.0 C Class null 18500 Automatic 28248 Diesel 2.1

In [0]:
%sql
/* first step to understanding unique values */
SELECT count(DISTINCT year) as dist_values_year
FROM cclass_csv
WHERE year IS NOT NULL

dist_values_year 22

> There are only `22` unique years in the dataset. The idea to create cluster with not null values, understand distribution and impute within cluster associated by other features (price, mileage, fuelType, engineSize). To re-create the same scenario, we take the first null year value, and compare with other not null value and analyze them.

> Check first null value
```
C Class
null
14477
Automatic
72000
Diesel
1.6
```

In [0]:
%sql
/* Checking the first null value above, we select only price between 12k and 15k, automatic transmission, diesel fuelType, 1.6 engineSize and mileage between 70k and 74k */
SELECT *
FROM cclass_csv
WHERE price BETWEEN 12000 AND 15000 AND
  transmission = 'Automatic' AND
  fuelType = 'Diesel' AND
  engineSize = '1.6' AND
  mileage BETWEEN 70000 AND 74000 AND
  year IS NOT NULL;

model year price transmission mileage fuelType engineSize

> Check with other wide range, maybe we haven't same characteristics like null value.

In [0]:
%sql
SELECT *
FROM cclass_csv
WHERE price BETWEEN 10000 AND 18000 AND
  transmission = 'Automatic' AND
  fuelType = 'Diesel' AND
  engineSize = '1.6' AND
  mileage BETWEEN 60000 AND 80000 AND
  year IS NOT NULL;

model year price transmission mileage fuelType engineSize C Class 2017-01-01 16500 Automatic 75000 Diesel 1.6

> And we have only one value with similar values, so in conclusion firt null value will be impute to `2017` ?

> With same approach, we analyze same procedure with the second null value, `mileage`.

In [0]:
%sql
/*# Check mileage null values cross table */
SELECT 
  mileage
  ,COUNT(*) as total_values
FROM cclass_csv
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20;

mileage total_values 1000 82 2000 21 16 20 3000 17 100 17 40005 17 45000 16 9 15 6000 14 20000 11 30676 11 54000 10 5000 10 4999 10 500 9 15000 9 3999 8 22000 8 4000 8 29000 8

In [0]:
%sql
/* check how many distinct values exists */
SELECT COUNT(DISTINCT mileage) as dist_values_mileage
FROM cclass_csv
WHERE mileage IS NOT NULL;

dist_values_mileage 3199

> There are only `3199` unique values for mileage feature. Of course, isn't discrete variable as year, but **continuous variable**, and so wide range to cluster or predict and impute null values.

In [0]:
%sql
/* # Check null values and all attributes for mileage feature */
SELECT *
FROM cclass_csv
WHERE mileage IS NULL;

model year price transmission mileage fuelType engineSize C Class 2020-01-01 30999 Automatic null Diesel 2.0 C Class 2015-01-01 16495 Automatic null Hybrid 2.0 C Class 2016-01-01 18498 Semi-Auto null Hybrid 2.0 C Class 2019-01-01 32998 Semi-Auto null Petrol 2.0 C Class 2015-01-01 13361 Semi-Auto null Petrol 2.0 C Class 2015-01-01 18299 Automatic null Diesel 2.1 C Class 2019-01-01 24995 Automatic null Petrol 1.5

> For instance, if we impute null values with median or mean, or we want to clusterize in groups to impute them with other techniques described above, regression for example, we could check per each observation distribution within group. Manually check and investigate the variables.

In [0]:
%sql
/* For the first null value above, we select only price between 30k and 32k, automatic transmission, diesel fuelType, 2.0 engineSize and year 2020 */
SELECT *
FROM cclass_csv
WHERE price BETWEEN 30000 AND 32000 AND
  transmission = 'Automatic' AND
  fuelType = 'Diesel' AND
  engineSize = '2.0' AND
  mileage IS NOT NULL AND
  YEAR(year) = 2020;

model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 31950 Automatic 2985 Diesel 2.0 C Class 2020-01-01 30990 Automatic 6000 Diesel 2.0 C Class 2020-01-01 31690 Automatic 3000 Diesel 2.0 C Class 2020-01-01 31980 Automatic 500 Diesel 2.0 C Class 2020-01-01 31590 Automatic 1452 Diesel 2.0 C Class 2020-01-01 30395 Automatic 2500 Diesel 2.0 C Class 2020-01-01 30480 Automatic 500 Diesel 2.0 C Class 2020-01-01 31900 Automatic 5000 Diesel 2.0

In [0]:
%sql
/* Visualize the price and mileage mean */
SELECT *
FROM cclass_csv
WHERE price BETWEEN 30000 AND 32000 AND
  transmission = 'Automatic' AND
  fuelType = 'Diesel' AND
  engineSize = '2.0' AND
  mileage IS NOT NULL AND
  YEAR(year) = 2020;

model year price transmission mileage fuelType engineSize C Class 2020-01-01 30495 Automatic 1200 Diesel 2.0 C Class 2020-01-01 31950 Automatic 2985 Diesel 2.0 C Class 2020-01-01 30990 Automatic 6000 Diesel 2.0 C Class 2020-01-01 31690 Automatic 3000 Diesel 2.0 C Class 2020-01-01 31980 Automatic 500 Diesel 2.0 C Class 2020-01-01 31590 Automatic 1452 Diesel 2.0 C Class 2020-01-01 30395 Automatic 2500 Diesel 2.0 C Class 2020-01-01 30480 Automatic 500 Diesel 2.0 C Class 2020-01-01 31900 Automatic 5000 Diesel 2.0

> We catch anormal distribution by mean of price, because probably depend the car features, and for instance, car model. In this dataset we have only feature represnt C-Class, not specified which kind C-Class model is.

***
Conclusion
> After checking year and mileage null values like steps described above to understanding distribution and take decision to remove them.

In [0]:
# Create a backup dataset
df_bak = df

In [0]:
# Check df_bak structure
df_bak.count()

Out[36]: 3899

In [0]:
# Drop Null values
df = df.dropna(how='any')

In [0]:
# Check after null values removed
df.count()

Out[38]: 3887

## 7. Feature Engineering

`Feature engineering` is about creating new input features from your existing ones.

In general, you can think of data cleaning as a process of subtraction and feature engineering as a process of addition.

This is often one of the most valuable tasks a data science area can do to improve model performance, for 3 big reasons:

- we can isolate and highlight key information, which helps your algorithms "focus" on what’s important.
- we can bring in your own domain expertise.
- most importantly, once you understand the "vocabulary" of feature engineering, you can bring in other people’s domain expertise!

Main tasks are :

* 1. combine sparse classes
* 2. add dummy variables
* 3. remove unused features

In [0]:
df.show()

+-------+----------+-----+------------+-------+--------+----------+
|  model|      year|price|transmission|mileage|fuelType|engineSize|
+-------+----------+-----+------------+-------+--------+----------+
|C Class|2020-01-01|30495|   Automatic|   1200|  Diesel|       2.0|
|C Class|2020-01-01|29989|   Automatic|   1000|  Petrol|       1.5|
|C Class|2020-01-01|37899|   Automatic|    500|  Diesel|       2.0|
|C Class|2019-01-01|30399|   Automatic|   5000|  Diesel|       2.0|
|C Class|2019-01-01|29899|   Automatic|   4500|  Diesel|       2.0|
|C Class|2020-01-01|35999|   Automatic|    500|  Diesel|       2.0|
|C Class|2019-01-01|37990|   Automatic|   1412|  Petrol|       3.0|
|C Class|2019-01-01|28990|   Automatic|   3569|  Diesel|       2.0|
|C Class|2019-01-01|28990|   Automatic|   3635|  Diesel|       2.0|
|C Class|2013-01-01| 9995|   Automatic|  44900|  Petrol|       1.6|
|C Class|2012-01-01| 6995|   Automatic|  88200|  Diesel|       2.1|
|C Class|2012-01-01| 7495|   Automatic| 115000| 

In [0]:
# Create new feature, carAge calculate the age of car.
from pyspark.sql.functions import year, datediff, date_format, current_date

In [0]:
df.show(5)

+-------+----------+-----+------------+-------+--------+----------+
|  model|      year|price|transmission|mileage|fuelType|engineSize|
+-------+----------+-----+------------+-------+--------+----------+
|C Class|2020-01-01|30495|   Automatic|   1200|  Diesel|       2.0|
|C Class|2020-01-01|29989|   Automatic|   1000|  Petrol|       1.5|
|C Class|2020-01-01|37899|   Automatic|    500|  Diesel|       2.0|
|C Class|2019-01-01|30399|   Automatic|   5000|  Diesel|       2.0|
|C Class|2019-01-01|29899|   Automatic|   4500|  Diesel|       2.0|
+-------+----------+-----+------------+-------+--------+----------+
only showing top 5 rows



In [0]:
# Remove null values
df.dropna(how='any', subset=['price','year'])

Out[42]: DataFrame[model: string, year: date, price: int, transmission: string, mileage: int, fuelType: string, engineSize: string]

In [0]:
# Remove columns year and model because aren't important features
df = df.drop('year','model')

In [0]:
df.count()

Out[44]: 3887

In [0]:
%sql
/* Show database and tables */
SHOW DATABASES;

databaseName default

In [0]:
%sql
SHOW TABLES;

database tableName isTemporary cclass_csv true

In [0]:
# Remove temporary parquet location
pqPath = 'dbfs:/user/hive/warehouse/cclass_cleaned'
dbutils.fs.rm(pqPath, True)

Out[45]: True

In [0]:
%sql
CREATE DATABASE pepito;

In [0]:
%sql
SHOW DATABASES;

databaseName default marco pepito

In [0]:
%sql
USE marco;

In [0]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.
permanent_table_name = "cclass_cleaned"

df.write \
  .format("parquet") \
  .mode("overwrite") \
  .saveAsTable(permanent_table_name)


In [0]:
%sql
SHOW TABLES;

database tableName isTemporary marco cclass_cleaned false cclass_csv true

In [0]:
%sql
SHOW TABLES;

database tableName isTemporary marco cclass_cleaned false cclass_csv true

In [0]:
%sql
SELECT * FROM cclass_cleaned LIMIT 3;

price transmission mileage fuelType engineSize 30495 Automatic 1200 Diesel 2.0 29989 Automatic 1000 Petrol 1.5 37899 Automatic 500 Diesel 2.0

## 8. Batch Inference - Delta Table

There are many scenarios where you might want to evaluate a model on a corpus of new data. For example, you may have a fresh batch of data, or may need to compare the performance of two models on the same corpus of data.

The following code evaluates the model on data stored in a Delta table, using Spark to run the computation in parallel.

In [0]:
# To simulate a new corpus of data, save the existing DF data to a Delta table. 
# In the real world, this would be a new batch of data.
#spark_df = spark.createDataFrame(df)

# Replace <username> with your username before running this cell.
table_path = "dbfs:/user/hive/warehouse/delta/cclass_cleaned"

# Delete the contents of this path in case this cell has already been run
dbutils.fs.rm(table_path, True)
df.write.format("delta").save(table_path)

> Load the model into a Spark UDF, so it can be applied to the Delta table.

In [0]:
# Read the "new data" from Delta
new_data = spark.read.format("delta").load(table_path)

In [0]:
# show results
display(new_data)

price transmission mileage fuelType engineSize 30495 Automatic 1200 Diesel 2.0 29989 Automatic 1000 Petrol 1.5 37899 Automatic 500 Diesel 2.0 30399 Automatic 5000 Diesel 2.0 29899 Automatic 4500 Diesel 2.0 35999 Automatic 500 Diesel 2.0 37990 Automatic 1412 Petrol 3.0 28990 Automatic 3569 Diesel 2.0 28990 Automatic 3635 Diesel 2.0 9995 Automatic 44900 Petrol 1.6 6995 Automatic 88200 Diesel 2.1 7495 Automatic 115000 Diesel 2.1 8995 Automatic 69250 Diesel 2.1 14995 Automatic 49850 Diesel 2.1 8595 Automatic 82685 Diesel 2.1 19250 Semi-Auto 32506 Diesel 2.1 23500 Semi-Auto 17000 Diesel 2.1 24995 Semi-Auto 17744 Diesel 2.0 29995 Semi-Auto 64291 Petrol 4.0 17990 Automatic 41011 Petrol 2.0 10977 Automatic 72000 Diesel 2.1 17977 Automatic 40000 Diesel 1.6 12790 Automatic 52000 Diesel 2.1 14488 Automatic 66472 Diesel 2.1 7500 Manual 91000 Diesel 2.1 9750 Automatic 69700 Diesel 2.1 15995 Automatic 39000 Diesel 1.6 15675 Manual 39287 Diesel 2.1 16475 Manual 40569 Diesel 2.1 15975 Manual 37419 Petrol 2.0 16475 Automatic 34452 Diesel 2.1 15975 Automatic 41567 Hybrid 2.1 17795 Automatic 37778 Diesel 2.1 17875 Automatic 44325 Diesel 2.1 14000 Automatic 45000 Diesel 2.1 16500 Automatic 75000 Diesel 1.6 21000 Automatic 36000 Diesel 2.1 15500 Automatic 64000 Diesel 2.1 7250 Manual 106000 Diesel 2.1 13000 Automatic 67150 Diesel 2.1 11000 Automatic 43426 Petrol 1.6 13300 Manual 42622 Diesel 2.1 16000 Manual 34903 Diesel 2.1 12500 Automatic 35345 Diesel 2.1 14495 Manual 23146 Petrol 2.0 24500 Automatic 16000 Diesel 2.1 23990 Automatic 44000 Petrol 6.2 17250 Automatic 30000 Diesel 2.1 14200 Automatic 22000 Petrol 1.6 22950 Automatic 19000 Petrol 2.0 10500 Automatic 90000 Diesel 2.1 15650 Automatic 32000 Diesel 2.1 14850 Automatic 90000 Diesel 2.1 22400 Automatic 37000 Diesel 2.1 19950 Automatic 23000 Diesel 2.1 9990 Automatic 78000 Diesel 2.1 11990 Automatic 47000 Diesel 2.1 20990 Automatic 18000 Diesel 2.1 14490 Automatic 64000 Diesel 2.1 10490 Automatic 50000 Diesel 2.1 15990 Automatic 65000 Diesel 2.1 17850 Automatic 38000 Diesel 2.1 16450 Automatic 70000 Diesel 2.1 11350 Automatic 69000 Diesel 2.1 12900 Automatic 41000 Petrol 1.6 26000 Automatic 16250 Petrol 1.5 22767 Automatic 30676 Diesel 2.1 12000 Automatic 40005 Diesel 2.1 18999 Automatic 38000 Hybrid 2.0 14000 Automatic 45000 Diesel 2.1 12000 Automatic 40005 Diesel 2.1 4343 Automatic 120000 Diesel 2.1 16500 Automatic 51936 Diesel 2.1 27750 Automatic 4000 Diesel 2.0 28950 Automatic 8500 Diesel 2.0 3495 Automatic 119168 Petrol 1.8 59740 Automatic 21800 Petrol 4.0 44990 Automatic 19999 Petrol 4.0 54990 Automatic 1600 Petrol 4.0 15995 Automatic 46221 Diesel 2.1 14000 Automatic 45000 Diesel 2.1 12000 Automatic 40005 Diesel 2.1 22767 Automatic 30676 Diesel 2.1 12998 Semi-Auto 52712 Diesel 2.1 17998 Semi-Auto 59858 Diesel 2.1 28998 Manual 22 Petrol 1.6 24998 Semi-Auto 28234 Petrol 2.0 19498 Semi-Auto 22803 Diesel 2.1 23498 Semi-Auto 17093 Diesel 2.1 25498 Semi-Auto 8150 Diesel 1.6 24998 Semi-Auto 9344 Diesel 2.0 15498 Semi-Auto 36243 Diesel 2.1 20995 Semi-Auto 33000 Diesel 2.1 31995 Semi-Auto 11300 Petrol 3.0 23995 Semi-Auto 32800 Diesel 2.1 16995 Automatic 37600 Diesel 2.1 24995 Automatic 14000 Petrol 2.0 26998 Semi-Auto 24175 Petrol 6.2 20498 Semi-Auto 27968 Diesel 2.1 27498 Semi-Auto 9255 Diesel 2.0 16998 Semi-Auto 26973 Diesel 1.6 27498 Semi-Auto 12491 Diesel 2.1 27698 Semi-Auto 14999 Diesel 2.0 27498 Semi-Auto 37941 Petrol 3.0 32498 Semi-Auto 11301 Diesel 2.0 24498 Semi-Auto 12629 Diesel 2.1 21498 Semi-Auto 20981 Diesel 2.1 18998 Semi-Auto 23760 Diesel 2.1 15998 Semi-Auto 26216 Diesel 2.1 9298 Semi-Auto 80485 Diesel 2.1 22998 Semi-Auto 18739 Diesel 2.1 12998 Manual 53453 Diesel 2.1 22998 Semi-Auto 23090 Diesel 2.1 17998 Manual 26100 Diesel 2.1 28498 Semi-Auto 4285 Diesel 2.0 18998 Semi-Auto 28388 Diesel 1.6 20998 Semi-Auto 16902 Diesel 2.1 21998 Semi-Auto 31227 Diesel 2.1 13998 Manual 24010 Diesel 1.6 24498 Semi-Auto 15838 Diesel 2.0 20998 Semi-Auto 17657 Hybrid 2.1 30498 Semi-Auto 4977 Dies

## 9. Summary

Test & Hypothesis

In order to understand what affects change in price of a used car, the relation between features available in the data sat will be examined by using inferential statistic methods. The primary assumption based on figures and tables is price must be affected by odometer and condition. There must be other features that affects price significantly. It will be investigated in the later phase of the study.

The first hypothesis:

Null Hypothesis: There is no significant relation between price and odometer of a car

Alt-Hypothesis: There is a significant relation between price and odometer.

The second hypothesis:

Null Hypothesis: There is no significant relation between price and condition of a car

Alt-Hypothesis: There is a significant relation between price and condition.

Odometer vs Price

Firstly, examine first hypothesis. Independent t-test is an appropriate method while examining the relation between two numerical variables. On the other hand, this test has some presumptions. Homogeneity of variances is one of the assumptions. In order to check whether homogeneity of variance assumption is violated or not, Levene test is applied.

## Check next notebook for ML Model: 

This section used applied machine learning models as a framework for the data analysis. The data set is a supervised data which refers to fitting a model of dependent variables to the independent variables, with the goal of accurately predicting the dependent variable for future observations or understanding the relationship between the variables.

- Pre-processing the data

- Label Encoding. In the dataset, there are 6 predictors. 3 of them are numerical variables while rest of them are categorical. In order to apply machine learning models, we need numeric representation of the features. Therefore, all non-numeric features were transformed into numerical form.

- Train the data. In this process, 20% of the data was split for the test data and 80% of the data was taken as train data.

- Scaling the Data. While exploring the data in the previous sections, it was seen that the data is not normally distributed. Without scaling, the machine learning models will try to disregard coefficients of features that has low values because their impact will be so small compared to the big value features.

**[Daimler]-DataScience_aptituteTest_MarcoRusso_v1.1.0_[part II]_python-scala**

## Version Control

- 1.2.0 - MR - 2021.07.20
```
AddDelta Table
Save Parquet Table
Remove check parquet table
```

- 1.1.0 
Original notebook